{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Saddleplots - plotting only !"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import standard python libraries\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import os, subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import python package for working with cooler files and tools for analysis\n",
    "import cooler\n",
    "import cooltools.lib.plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "# from saddle import saddleplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download test data\n",
    "# this file is 145 Mb, and may take a few seconds to download\n",
    "import bbi\n",
    "import cooltools\n",
    "import bioframe\n",
    "from matplotlib.colors import LogNorm\n",
    "from helper_func import saddleplot\n",
    "from data_catalog import bws, bws_vlim, telo_dict\n",
    "\n",
    "import saddle\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "from tqdm.notebook import trange, tqdm\n",
    "import warnings\n",
    "import seaborn as sns\n",
    "\n",
    "import warnings\n",
    "import h5py\n",
    "\n",
    "import matplotlib.lines as lines\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.patches import ConnectionPatch, Rectangle\n",
    "from mpl_toolkits.axes_grid1 import Divider, Size\n",
    "from mpl_toolkits.axes_grid1.inset_locator import BboxConnector\n",
    "from matplotlib import cm\n",
    "import matplotlib as mpl\n",
    "# from mpl_toolkits.axes_grid1.Size import Fixed\n",
    "\n",
    "\n",
    "# enable editable text ...\n",
    "mpl.rcParams[\"pdf.fonttype\"]=42\n",
    "mpl.rcParams[\"svg.fonttype\"]=\"none\"\n",
    "mpl.rcParams['axes.linewidth'] = 0.5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating per-chromosome compartmentalization\n",
    "\n",
    "We first load the Hi-C data at 100 kbp resolution. \n",
    "\n",
    "Note that the current implementation of eigendecomposition in cooltools assumes that individual regions can be held in memory-- for hg38 at 100kb this is either a 2422x2422 matrix for chr2, or a 3255x3255 matrix for the full cooler here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define genomic view that will be used to call dots and pre-compute expected\n",
    "\n",
    "# Use bioframe to fetch the genomic features from the UCSC.\n",
    "hg38_chromsizes = bioframe.fetch_chromsizes('hg38')\n",
    "hg38_cens = bioframe.fetch_centromeres('hg38')\n",
    "hg38_arms_full = bioframe.make_chromarms(hg38_chromsizes, hg38_cens)\n",
    "# # remove \"bad\" chromosomes and near-empty arms ...\n",
    "included_arms = hg38_arms_full[\"name\"].to_list()[:44] # all autosomal ones ...\n",
    "hg38_arms = hg38_arms_full[hg38_arms_full[\"name\"].isin(included_arms)].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cre_to_code = {\n",
    "    'pls':6,\n",
    "    'pels':5,\n",
    "    'nels':5,\n",
    "    'dels':4,\n",
    "    'openk4':3,\n",
    "    'ctcf':2,\n",
    "    'justopen':1,\n",
    "}\n",
    "\n",
    "_cats = pd.CategoricalDtype(categories=list(range(6+1)), ordered=True)\n",
    "\n",
    "cre_fnames = {\n",
    "    \"atac95_rest95\" : \"./bigbed/all_cres_encodestyle_pMunion_95atac_95rest.bed\",  # control sorta, like before\n",
    "    # \"atac90_rest90\" : \"./bigbed/all_cres_encodestyle_pMunion_90atac_90rest.bed\",\n",
    "    # \"atac70_rest70\" : \"./bigbed/all_cres_encodestyle_pMunion_70atac_70rest.bed\",\n",
    "    # \"atac70_rest95\" : \"./bigbed/all_cres_encodestyle_pMunion_70atac_95rest.bed\",\n",
    "    # \"atac70_rest90\" : \"./bigbed/all_cres_encodestyle_pMunion_70atac_90rest.bed\",\n",
    "    # \"atac75_rest95\" : \"./bigbed/all_cres_encodestyle_pMunion_75atac_95rest.bed\",\n",
    "    # \"atac75_rest90\" : \"./bigbed/all_cres_encodestyle_pMunion_75atac_90rest.bed\",\n",
    "    # \"allana_latest\" : \"./bigbed/all_cres_encodestyle_pMunion_NEWatac_k2785_rest95.bed\",\n",
    "}\n",
    "\n",
    "_status_query = \"(status_ == 0)|(status_ == 5)|(status_ == 4)|(status_ == 6)|(status_ == 7)\"\n",
    "\n",
    "binned_cre10_dfs = {}\n",
    "binned_cre25_dfs = {}\n",
    "for name, cre_fname in cre_fnames.items():\n",
    "    df = pd.read_table(cre_fname, usecols=['chrom', 'start', 'end', 'name'])\n",
    "    print(f\"read {name=} with {len(df)} items ...\")\n",
    "    # # rename nels -> pels\n",
    "    # df = df.replace({'name': {\"nels\": \"pels\"}})\n",
    "    # sort by coordinate\n",
    "    df = bioframe.sort_bedframe(df, view_df=hg38_arms)\n",
    "    df[\"status\"] = df[\"name\"].map(cre_to_code)\n",
    "\n",
    "    # # now annotate bins with the anchors @10kb  ...\n",
    "    # _bin_cres = bioframe.overlap(clr_bins10, df)\n",
    "    # _bin_cres = _bin_cres.drop(columns=[\"chrom_\",\"start_\",\"end_\"])\n",
    "    # _bin_cres[\"status_\"] = _bin_cres[\"status_\"].fillna(0).astype(int)\n",
    "    # # assign cCREs per bin according to the hierarchy ...\n",
    "    # binned_cres = _bin_cres.groupby([\"chrom\",\"start\",\"end\"], observed=True)[\"status_\"].max().reset_index()\n",
    "    # binned_cres[\"status_\"] = binned_cres[\"status_\"].astype(_cats)\n",
    "    # # value_counts of annotated bins before gene assignment ...\n",
    "    # print(\"@10kb\\n\")\n",
    "    # display(binned_cres[\"status_\"].value_counts().sort_index())\n",
    "    # print(\"\\n\")\n",
    "    # #\n",
    "    # binned_cre10_dfs[name] = binned_cres.astype({ \"chrom\" : str })\n",
    "\n",
    "    # # now annotate bins with the anchors @25kb  ...\n",
    "    # _bin_cres = bioframe.overlap(clr_bins25, df)\n",
    "    # _bin_cres = _bin_cres.drop(columns=[\"chrom_\",\"start_\",\"end_\"])\n",
    "    # _bin_cres[\"status_\"] = _bin_cres[\"status_\"].fillna(0).astype(int)\n",
    "    # # assign cCREs per bin according to the hierarchy ...\n",
    "    # binned_cres = _bin_cres.groupby([\"chrom\",\"start\",\"end\"], observed=True)[\"status_\"].max().reset_index()\n",
    "    # binned_cres[\"status_\"] = binned_cres[\"status_\"].astype(_cats)\n",
    "    # # value_counts of annotated bins before gene assignment ...\n",
    "    # print(\"@25kb\\n\")\n",
    "    # display(binned_cres[\"status_\"].value_counts().sort_index())\n",
    "    # print(\"\\n\")\n",
    "    # # ...\n",
    "    # binned_cre25_dfs[name] = binned_cres.astype({ \"chrom\" : str })\n",
    "\n",
    "\n",
    "\n",
    "ticklabels = [\n",
    "    \"PLS\",\n",
    "    \"pELS\",\n",
    "    \"dELS\",\n",
    "    \"opK4\",\n",
    "    \"CTCF\",\n",
    "    \"open\",\n",
    "    \"none\",\n",
    "]\n",
    "# # # categroies are going to be 1,2,3,4,5,6\n",
    "# # # 0 - everything else ...\n",
    "\n",
    "# _working_cres10 = binned_cre10_dfs['atac95_rest95']\n",
    "# _working_cres25 = binned_cre25_dfs['atac95_rest95']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# fr.items()\n",
    "def print_attrs(name, obj):\n",
    "    # Create indent\n",
    "    shift = name.count('/') * '    '\n",
    "    item_name = name.split(\"/\")[-1]\n",
    "    print(shift + item_name)\n",
    "    try:\n",
    "        for key, val in obj.attrs.items():\n",
    "            print(shift + '    ' + f\"{key}: {val}\")\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "\n",
    "with h5py.File(\"saddles_cre_by_distance.hdf5\", 'r') as fr:\n",
    "# with h5py.File(\"saddles_cre_by_distance_Allana_latest_cre.hdf5\", 'r') as fr:\n",
    "    # f = h5py.File('foo.hdf5','r')\n",
    "    fr.visititems(print_attrs)\n",
    "\n",
    "    # check general metadata ...\n",
    "    _saddle_meta = dict(fr.attrs)\n",
    "\n",
    "    # sort out the results ...\n",
    "    interaction_sums_trans = {}\n",
    "    interaction_counts_trans = {}\n",
    "    # sort out the results ...\n",
    "    _counts = fr.get(\"counts_trans\")\n",
    "    for _sample in _counts.keys():\n",
    "        interaction_counts_trans[_sample] = _counts.get(_sample)[()]\n",
    "    _sums = fr.get(\"sums_trans\")\n",
    "    for _sample in _sums.keys():\n",
    "        interaction_sums_trans[_sample] = _sums.get(_sample)[()]\n",
    "\n",
    "    interaction_sums = {}\n",
    "    interaction_counts = {}\n",
    "    # sort out the results ...\n",
    "    _counts = fr.get(\"counts\")\n",
    "    for _sample in _counts.keys():\n",
    "        interaction_counts[_sample] = _counts.get(_sample)[()]\n",
    "    _sums = fr.get(\"sums\")\n",
    "    for _sample in _sums.keys():\n",
    "        interaction_sums[_sample] = _sums.get(_sample)[()]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_samples_m = [\n",
    "    \"mMito\",\n",
    "    \"mTelo\",\n",
    "    \"mCyto\",\n",
    "    \"m5hR1R2\",\n",
    "    \"m10hR1R2\",\n",
    "]\n",
    "sub_samples_p = [\n",
    "    \"pMito\",\n",
    "    \"pTelo\",\n",
    "    \"pCyto\",\n",
    "    \"p5hR1R2\",\n",
    "    \"p10hR1R2\",\n",
    "]\n",
    "\n",
    "\n",
    "# introduce distance ranges\n",
    "distances = {\n",
    "    \"<0.25MB\": slice(0,int(250_000/_saddle_meta[\"cis_binsize\"])+1),\n",
    "    \"0.25-1Mb\": slice(int(250_000/_saddle_meta[\"cis_binsize\"]),int(1_000_000/_saddle_meta[\"cis_binsize\"])+1),\n",
    "    \"1-10Mb\": slice(int(1_000_000/_saddle_meta[\"cis_binsize\"]),int(10_000_000/_saddle_meta[\"cis_binsize\"])+1),\n",
    "    \">10Mb\": slice(int(10_000_000/_saddle_meta[\"cis_binsize\"]),int(1_000_000_000/_saddle_meta[\"cis_binsize\"])+1),\n",
    "    \"trans\": slice(None),\n",
    "}\n",
    "\n",
    "\n",
    "imshow_kwargs = dict(\n",
    "        norm=LogNorm(vmin=1/2, vmax=2),\n",
    "        cmap=\"RdBu_r\",\n",
    "        interpolation=\"nearest\",\n",
    ")\n",
    "\n",
    "legend_kwargs = dict(\n",
    "    aspect=\"auto\",\n",
    "    interpolation=\"nearest\",\n",
    ")\n",
    "\n",
    "cre_cmap = {\n",
    "    6: \"darkred\",  #\"PLS\",\n",
    "    5: \"red\",  #\"pELS\",\n",
    "    4: \"orangered\",  #\"dELS\",\n",
    "    3: \"pink\",  #\"opK4\",\n",
    "    2: \"blue\",  #\"CTCF\",\n",
    "    1: \"#ffee99\",  #\"open\",\n",
    "    0: \"#D9E2EF\",  #\"none\"\n",
    "}\n",
    "\n",
    "# some common sizings for the saddles ...\n",
    "# cbarw = 0.7*matw\n",
    "margin = 0.2\n",
    "matw = 0.6\n",
    "cbarh = 0.08\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Draw actual figures with the semi-manual custom layout ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The first items are for padding and the second items are for the axes, sizes are in inch.\n",
    "h = [ Size.Fixed(margin) ] + len(distances)*[Size.Fixed(matw), Size.Fixed(0.2*margin)] + [Size.Fixed(cbarh), Size.Fixed(margin)]\n",
    "# goes from bottom to the top ...\n",
    "v = [Size.Fixed(margin), Size.Fixed(cbarh)] + 2 * [Size.Fixed(0.2*margin), Size.Fixed(matw)] + [Size.Fixed(0.2*margin), Size.Fixed(cbarh)]\n",
    "\n",
    "_stickingout_bit = sum(_h.fixed_size for _h in h[-3:])\n",
    "print(f\"{_stickingout_bit=}\")\n",
    "\n",
    "\n",
    "\n",
    "for _fig_fname, (sample_m, sample_p) in {\n",
    "    \"Fig2F.svg\": (\"m5hR1R2\", \"p5hR1R2\"),\n",
    "    \"FigE2D.svg\": (\"N93m5\", \"N93p5\"),\n",
    "}.items():\n",
    "    # set figsize based on the tiling provided ...\n",
    "    fig_width = sum(_h.fixed_size for _h in h)\n",
    "    fig_height = sum(_v.fixed_size for _v in v)\n",
    "    fig = plt.figure(\n",
    "        figsize=(fig_width, fig_height),\n",
    "        # facecolor='lightblue'\n",
    "    )\n",
    "    print(f\"figure size {fig_width=} {fig_height=}\")\n",
    "\n",
    "    # ...\n",
    "    divider = Divider(fig, (0, 0, 1, 1), h, v, aspect=False)\n",
    "\n",
    "    axs_m = {}\n",
    "    axs_p = {}\n",
    "    cax_h = {}\n",
    "\n",
    "    for i, _dist_key in enumerate(distances):\n",
    "        # mind the gaps/marging between actual plots ...\n",
    "        axs_p[_dist_key] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*i+1, ny=3))\n",
    "        axs_m[_dist_key] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*i+1, ny=5))\n",
    "        # ...\n",
    "        cax_h[_dist_key] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*i+1, ny=7))\n",
    "\n",
    "    caxp_v = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*(i+1)+1, ny=3))\n",
    "    caxm_v = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*(i+1)+1, ny=5))\n",
    "\n",
    "    cbar_ax = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*i+1, ny=1))\n",
    "\n",
    "    for _ax in list(cax_h.values()) + [caxm_v, caxp_v, cbar_ax]:\n",
    "        _ax.set_xticks([])\n",
    "        _ax.set_yticks([])\n",
    "\n",
    "    for jj, (_dist_key, _dist) in enumerate(distances.items()):\n",
    "        axm, axp, cax = axs_m[_dist_key], axs_p[_dist_key], cax_h[_dist_key]\n",
    "        if _dist_key != \"trans\":\n",
    "            Cm = np.nanmean(interaction_sums[sample_m][_dist], axis=0) / np.nanmean(interaction_counts[sample_m][_dist], axis=0)\n",
    "            Cp = np.nanmean(interaction_sums[sample_p][_dist], axis=0) / np.nanmean(interaction_counts[sample_p][_dist], axis=0)\n",
    "        elif _dist_key == \"trans\":\n",
    "            # pass\n",
    "            Cm = np.nanmean(interaction_sums_trans[sample_m][_dist], axis=0) / np.nanmean(interaction_counts_trans[sample_m][_dist], axis=0)\n",
    "            Cp = np.nanmean(interaction_sums_trans[sample_p][_dist], axis=0) / np.nanmean(interaction_counts_trans[sample_p][_dist], axis=0)\n",
    "        else:\n",
    "            pass\n",
    "        _hm = axm.imshow(Cm, **imshow_kwargs)\n",
    "        _hm.cmap.set_over(\"#300000\")\n",
    "        _hm = axp.imshow(Cp, **imshow_kwargs)\n",
    "        _hm.cmap.set_over(\"#300000\")\n",
    "\n",
    "        for _ax in [axp, axm]:\n",
    "            _ax.set_xticks([])\n",
    "            _ax.set_yticks([])\n",
    "        # axm.set_title(f\"{_dist_key}\", fontsize=9)\n",
    "        cax.set_title(f\"{_dist_key}\", fontsize=8)\n",
    "        # axp.set_xticks(np.arange(len(ticklabels)))\n",
    "        # axp.set_xticklabels(np.asarray(ticklabels[::-1]), rotation=\"vertical\")\n",
    "        if jj == 0:\n",
    "            axm.set_ylabel(sample_m, fontsize=8)\n",
    "            axp.set_ylabel(sample_p, fontsize=8)\n",
    "\n",
    "    # create a \"fake\" legend for CREs - for now ...\n",
    "    _size, _ = Cm.shape\n",
    "    # ...\n",
    "    _fdata = np.reshape(np.arange(_size), (-1,1))\n",
    "    _fcmap = plt.cm.gray\n",
    "    _fnorm = plt.Normalize()\n",
    "    _frgba = _fcmap(_fnorm(_fdata))\n",
    "    _frgbaT = _fcmap(_fnorm(_fdata.T))\n",
    "    # ...\n",
    "    # Set the diagonal to red\n",
    "    for i in range(_size):\n",
    "        _frgba[i,0] = list(mpl.colors.to_rgb(cre_cmap[i]))+[1]\n",
    "        _frgbaT[0,i] = list(mpl.colors.to_rgb(cre_cmap[i]))+[1]\n",
    "\n",
    "    caxp_v.imshow(_frgba, **legend_kwargs)\n",
    "    caxm_v.imshow(_frgba, **legend_kwargs)\n",
    "\n",
    "    for _dist_key in distances:\n",
    "        cax_h[_dist_key].imshow(_frgbaT, **legend_kwargs)\n",
    "\n",
    "    # add a single colorbar ...\n",
    "\n",
    "    fig.colorbar(\n",
    "        cm.ScalarMappable(norm=imshow_kwargs[\"norm\"], cmap=imshow_kwargs[\"cmap\"]),\n",
    "        cax=cbar_ax,\n",
    "        orientation=\"horizontal\",\n",
    "    )\n",
    "    cbar_ax.set_xticks([imshow_kwargs[\"norm\"].vmin, 1, imshow_kwargs[\"norm\"].vmax])\n",
    "    cbar_ax.set_xticklabels([imshow_kwargs[\"norm\"].vmin, 1, imshow_kwargs[\"norm\"].vmax], fontsize=6)\n",
    "    cbar_ax.minorticks_off()\n",
    "    cbar_ax.tick_params(length=1.5, pad=1)#,direction='out', length=6, width=2, colors='r', grid_color='r', grid_alpha=0.5)\n",
    "    for _tidx, tick in enumerate(cbar_ax.xaxis.get_majorticklabels()):\n",
    "        if _tidx == 0:\n",
    "            tick.set_horizontalalignment(\"left\")\n",
    "        elif _tidx == 2:\n",
    "            tick.set_horizontalalignment(\"right\")\n",
    "        else:\n",
    "            tick.set_horizontalalignment(\"center\")\n",
    "\n",
    "    # # # # plt.show()\n",
    "    fig.savefig(_fig_fname, dpi=300)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Do a similar thing for the extended figure 3 ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_m = [\n",
    "    \"mMito\",\n",
    "    \"mTelo\",\n",
    "    \"mCyto\",\n",
    "    \"m5hR1R2\",\n",
    "]\n",
    "samples_p = [\n",
    "    \"pMito\",\n",
    "    \"pTelo\",\n",
    "    \"pCyto\",\n",
    "    \"p5hR1R2\",\n",
    "]\n",
    "assert len(samples_m) == len(samples_p)\n",
    "_num_timepoints = len(samples_m)\n",
    "\n",
    "# The first items are for padding and the second items are for the axes, sizes are in inch.\n",
    "h = [ Size.Fixed(margin) ] + \\\n",
    "    (len(distances) - 1)*[Size.Fixed(matw), Size.Fixed(0.2*margin)] + [Size.Fixed(matw), Size.Fixed(margin)] + \\\n",
    "    len(distances)*[Size.Fixed(matw), Size.Fixed(0.2*margin)] + \\\n",
    "    [Size.Fixed(cbarh), Size.Fixed(margin)]\n",
    "# goes from bottom to the top ...\n",
    "v = [Size.Fixed(margin), Size.Fixed(cbarh)] + \\\n",
    "    _num_timepoints*[Size.Fixed(0.2*margin), Size.Fixed(matw)] + \\\n",
    "    [Size.Fixed(0.2*margin), Size.Fixed(cbarh)] + \\\n",
    "    [Size.Fixed(margin)]\n",
    "\n",
    "_stickingout_bit = sum(_h.fixed_size for _h in h[-3:])\n",
    "print(f\"{_stickingout_bit=}\")\n",
    "\n",
    "# set figsize based on the tiling provided ...\n",
    "fig_width = sum(_h.fixed_size for _h in h)\n",
    "fig_height = sum(_v.fixed_size for _v in v)\n",
    "fig = plt.figure(\n",
    "    figsize=(fig_width, fig_height),\n",
    "    # facecolor='lightblue'\n",
    ")\n",
    "print(f\"figure size {fig_width=} {fig_height=}\")\n",
    "\n",
    "# ...\n",
    "divider = Divider(fig, (0, 0, 1, 1), h, v, aspect=False)\n",
    "\n",
    "axs_p = {}\n",
    "axs_m = {}\n",
    "axl_m = {}\n",
    "axl_p = {}\n",
    "axlv_m = {}\n",
    "\n",
    "for i, _dist_key in enumerate(distances):\n",
    "    _nxm = 2*i+1\n",
    "    _nxp = 2*(i+len(distances))+1\n",
    "    axs_p[_dist_key] = {}\n",
    "    axs_m[_dist_key] = {}\n",
    "    for j, (_m, _p) in enumerate(zip(reversed(samples_m), reversed(samples_p))):\n",
    "        # mind the gaps/marging between actual plots ...\n",
    "        _ny = 2*j+3\n",
    "        axs_m[_dist_key][_m] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=_nxm, ny=_ny))\n",
    "        axs_p[_dist_key][_p] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=_nxp, ny=_ny))\n",
    "    # the last j+1 for the label on top ...\n",
    "    _ytop = 2*(j+1)+3\n",
    "    axl_m[_dist_key] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=_nxm, ny=_ytop))\n",
    "    axl_p[_dist_key] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=_nxp, ny=_ytop))\n",
    "\n",
    "_nxright = 4*len(distances)+1\n",
    "for j, _p in enumerate(reversed(samples_p)):\n",
    "    # mind the gaps/marging between actual plots ...\n",
    "    _ny = 2*j+3\n",
    "    axlv_m[_p] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=_nxright, ny=_ny))\n",
    "\n",
    "# separate colorbar axes\n",
    "cbar_ax = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=_nxright-2, ny=1))\n",
    "\n",
    "# turn of ticks everywhere ... for saddles first\n",
    "# for label axes and cbar afterwards ...\n",
    "for ax in (\n",
    "    sum([list(v.values()) for k,v in axs_p.items()], start=[]) + sum([list(v.values()) for k,v in axs_m.items()], start=[]) +\n",
    "    list(axl_m.values()) + list(axl_p.values()) + list(axlv_m.values()) + [cbar_ax]\n",
    "):\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "\n",
    "for _dist_key, _dist in distances.items():\n",
    "    for _m, _p in zip(samples_m, samples_p):\n",
    "        axm = axs_m[_dist_key][_m]\n",
    "        axp = axs_p[_dist_key][_p]\n",
    "        if _dist_key != \"trans\":\n",
    "            Cm = np.nanmean(interaction_sums[_m][_dist], axis=0) / np.nanmean(interaction_counts[_m][_dist], axis=0)\n",
    "            Cp = np.nanmean(interaction_sums[_p][_dist], axis=0) / np.nanmean(interaction_counts[_p][_dist], axis=0)\n",
    "        elif _dist_key == \"trans\":\n",
    "            Cm = np.nanmean(interaction_sums_trans[_m][_dist], axis=0) / np.nanmean(interaction_counts_trans[_m][_dist], axis=0)\n",
    "            Cp = np.nanmean(interaction_sums_trans[_p][_dist], axis=0) / np.nanmean(interaction_counts_trans[_p][_dist], axis=0)\n",
    "        else:\n",
    "            pass\n",
    "        # plot corresponding saddle ...\n",
    "        _hm = axm.imshow(Cm, **imshow_kwargs)\n",
    "        _hm.cmap.set_over(\"#300000\")\n",
    "        _hm = axp.imshow(Cp, **imshow_kwargs)\n",
    "        _hm.cmap.set_over(\"#300000\")\n",
    "        if _dist_key.startswith(\"<\"):\n",
    "            axm.set_ylabel(_m.lstrip(\"m\"), fontsize=8, labelpad=1)\n",
    "            # axp.set_ylabel(_p, fontsize=8, labelpad=1)\n",
    "    axl_m[_dist_key].set_title(f\"{_dist_key}\", fontsize=8, pad=2)\n",
    "    axl_p[_dist_key].set_title(f\"{_dist_key}\", fontsize=8, pad=2)\n",
    "\n",
    "# create a \"fake\" legend for CREs - for now ...\n",
    "_size, _ = Cm.shape\n",
    "# ...\n",
    "_fdata = np.reshape(np.arange(_size), (-1,1))\n",
    "_fcmap = plt.cm.gray\n",
    "_fnorm = plt.Normalize()\n",
    "_frgba = _fcmap(_fnorm(_fdata))\n",
    "_frgbaT = _fcmap(_fnorm(_fdata.T))\n",
    "# ...\n",
    "# Set the diagonal to red\n",
    "for i in range(_size):\n",
    "    _frgba[i,0] = list(mpl.colors.to_rgb(cre_cmap[i]))+[1]\n",
    "    _frgbaT[0,i] = list(mpl.colors.to_rgb(cre_cmap[i]))+[1]\n",
    "\n",
    "for _dist_key in distances:\n",
    "    axl_m[_dist_key].imshow(_frgbaT, **legend_kwargs)\n",
    "    axl_p[_dist_key].imshow(_frgbaT, **legend_kwargs)\n",
    "for _p in samples_p:\n",
    "    axlv_m[_p].imshow(_frgba, **legend_kwargs)\n",
    "\n",
    "# try adding labels to the right ...\n",
    "axlv_m[samples_p[-1]].yaxis.tick_right()\n",
    "axlv_m[samples_p[-1]].set_yticks(\n",
    "    list(cre_cmap.keys()),\n",
    "    labels=ticklabels,\n",
    "    fontsize=6,\n",
    ")\n",
    "axlv_m[samples_p[-1]].tick_params(length=1.5, pad=1)\n",
    "\n",
    "# add a single colorbar ...\n",
    "fig.colorbar(\n",
    "    cm.ScalarMappable(norm=imshow_kwargs[\"norm\"], cmap=imshow_kwargs[\"cmap\"]),\n",
    "    cax=cbar_ax,\n",
    "    orientation=\"horizontal\",\n",
    ")\n",
    "cbar_ax.set_xticks([imshow_kwargs[\"norm\"].vmin, 1, imshow_kwargs[\"norm\"].vmax])\n",
    "cbar_ax.set_xticklabels([imshow_kwargs[\"norm\"].vmin, 1, imshow_kwargs[\"norm\"].vmax], fontsize=6)\n",
    "cbar_ax.minorticks_off()\n",
    "cbar_ax.tick_params(length=1.5, pad=1)\n",
    "for _tidx, tick in enumerate(cbar_ax.xaxis.get_majorticklabels()):\n",
    "    if _tidx == 0:\n",
    "        tick.set_horizontalalignment(\"left\")\n",
    "    elif _tidx == 2:\n",
    "        tick.set_horizontalalignment(\"right\")\n",
    "    else:\n",
    "        tick.set_horizontalalignment(\"center\")\n",
    "\n",
    "\n",
    "_fig_fname = \"FigExt3B.svg\"\n",
    "fig.savefig(_fig_fname, dpi=300)\n",
    "\n",
    "! cairosvg --format pdf -o FigExt3B.pdf FigExt3B.svg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Draw Extended fig 7 or so - the 10hour mp experiment ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The first items are for padding and the second items are for the axes, sizes are in inch.\n",
    "h = [ Size.Fixed(margin) ] + len(distances)*[Size.Fixed(matw), Size.Fixed(0.2*margin)] + [Size.Fixed(cbarh), Size.Fixed(margin)]\n",
    "# goes from bottom to the top ...\n",
    "v = [Size.Fixed(margin), Size.Fixed(cbarh)] + 3 * [Size.Fixed(0.2*margin), Size.Fixed(matw)] + [Size.Fixed(0.2*margin), Size.Fixed(cbarh), Size.Fixed(margin)]\n",
    "\n",
    "_stickingout_bit = sum(_h.fixed_size for _h in h[-3:])\n",
    "print(f\"{_stickingout_bit=}\")\n",
    "\n",
    "sample_m, sample_mp, sample_p = \"m10hR1R2\", \"mp10hR1R2\", \"p10hR1R2\"\n",
    "# set figsize based on the tiling provided ...\n",
    "fig_width = sum(_h.fixed_size for _h in h)\n",
    "fig_height = sum(_v.fixed_size for _v in v)\n",
    "fig = plt.figure(\n",
    "    figsize=(fig_width, fig_height),\n",
    "    # facecolor='lightblue'\n",
    ")\n",
    "print(f\"figure size {fig_width=} {fig_height=}\")\n",
    "\n",
    "# ...\n",
    "divider = Divider(fig, (0, 0, 1, 1), h, v, aspect=False)\n",
    "\n",
    "axs_m = {}\n",
    "axs_mp = {}\n",
    "axs_p = {}\n",
    "cax_h = {}\n",
    "\n",
    "for i, _dist_key in enumerate(distances):\n",
    "    # mind the gaps/marging between actual plots ...\n",
    "    axs_m[_dist_key] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*i+1, ny=7))\n",
    "    axs_mp[_dist_key] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*i+1, ny=5))\n",
    "    axs_p[_dist_key] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*i+1, ny=3))\n",
    "    # ...\n",
    "    cax_h[_dist_key] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*i+1, ny=9))\n",
    "\n",
    "caxm_v = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*(i+1)+1, ny=7))\n",
    "caxmp_v = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*(i+1)+1, ny=5))\n",
    "caxp_v = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*(i+1)+1, ny=3))\n",
    "cbar_ax = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=2*i+1, ny=1))\n",
    "\n",
    "for ax in list(cax_h.values()) + [caxm_v, caxmp_v, caxp_v, cbar_ax]:\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "for jj, (_dist_key, _dist) in enumerate(distances.items()):\n",
    "    axm, axmp, axp, cax = axs_m[_dist_key], axs_mp[_dist_key], axs_p[_dist_key], cax_h[_dist_key]\n",
    "    if _dist_key != \"trans\":\n",
    "        Cm = np.nanmean(interaction_sums[sample_m][_dist], axis=0) / np.nanmean(interaction_counts[sample_m][_dist], axis=0)\n",
    "        Cmp = np.nanmean(interaction_sums[sample_mp][_dist], axis=0) / np.nanmean(interaction_counts[sample_mp][_dist], axis=0)\n",
    "        Cp = np.nanmean(interaction_sums[sample_p][_dist], axis=0) / np.nanmean(interaction_counts[sample_p][_dist], axis=0)\n",
    "    elif _dist_key == \"trans\":\n",
    "        Cm = np.nanmean(interaction_sums_trans[sample_m][_dist], axis=0) / np.nanmean(interaction_counts_trans[sample_m][_dist], axis=0)\n",
    "        Cmp = np.nanmean(interaction_sums_trans[sample_mp][_dist], axis=0) / np.nanmean(interaction_counts_trans[sample_mp][_dist], axis=0)\n",
    "        Cp = np.nanmean(interaction_sums_trans[sample_p][_dist], axis=0) / np.nanmean(interaction_counts_trans[sample_p][_dist], axis=0)\n",
    "    else:\n",
    "        pass\n",
    "    _hm = axm.imshow(Cm, **imshow_kwargs)\n",
    "    _hm.cmap.set_over(\"#300000\")\n",
    "    _hm = axmp.imshow(Cmp, **imshow_kwargs)\n",
    "    _hm.cmap.set_over(\"#300000\")\n",
    "    _hm = axp.imshow(Cp, **imshow_kwargs)\n",
    "    _hm.cmap.set_over(\"#300000\")\n",
    "    for _ax in [axp, axmp, axm]:\n",
    "        _ax.set_xticks([])\n",
    "        _ax.set_yticks([])\n",
    "    cax.set_title(f\"{_dist_key}\", fontsize=8)\n",
    "    if jj == 0:\n",
    "        axm.set_ylabel(sample_m.rstrip(\"R1R2\"), fontsize=8, labelpad=1)\n",
    "        axmp.set_ylabel(sample_mp.rstrip(\"R1R2\"), fontsize=8, labelpad=1)\n",
    "        axp.set_ylabel(sample_p.rstrip(\"R1R2\"), fontsize=8, labelpad=1)\n",
    "\n",
    "# create a \"fake\" legend for CREs - for now ...\n",
    "_size, _ = Cm.shape\n",
    "# ...\n",
    "_fdata = np.reshape(np.arange(_size), (-1,1))\n",
    "_fcmap = plt.cm.gray\n",
    "_fnorm = plt.Normalize()\n",
    "_frgba = _fcmap(_fnorm(_fdata))\n",
    "_frgbaT = _fcmap(_fnorm(_fdata.T))\n",
    "# ...\n",
    "# Set the diagonal to red\n",
    "for i in range(_size):\n",
    "    _frgba[i,0] = list(mpl.colors.to_rgb(cre_cmap[i]))+[1]\n",
    "    _frgbaT[0,i] = list(mpl.colors.to_rgb(cre_cmap[i]))+[1]\n",
    "\n",
    "caxp_v.imshow(_frgba, **legend_kwargs)\n",
    "caxm_v.imshow(_frgba, **legend_kwargs)\n",
    "caxmp_v.imshow(_frgba, **legend_kwargs)\n",
    "\n",
    "for _dist_key in distances:\n",
    "    cax_h[_dist_key].imshow(_frgbaT, **legend_kwargs)\n",
    "\n",
    "\n",
    "# try adding labels to the right ...\n",
    "caxp_v.yaxis.tick_right()\n",
    "caxp_v.set_yticks(\n",
    "    list(cre_cmap.keys()),\n",
    "    labels=ticklabels,\n",
    "    fontsize=6,\n",
    ")\n",
    "caxp_v.tick_params(length=1.5, pad=1)\n",
    "\n",
    "\n",
    "# add a single colorbar ...\n",
    "fig.colorbar(\n",
    "    cm.ScalarMappable(norm=imshow_kwargs[\"norm\"], cmap=imshow_kwargs[\"cmap\"]),\n",
    "    cax=cbar_ax,\n",
    "    orientation=\"horizontal\",\n",
    ")\n",
    "cbar_ax.set_xticks([imshow_kwargs[\"norm\"].vmin, 1, imshow_kwargs[\"norm\"].vmax])\n",
    "cbar_ax.set_xticklabels([imshow_kwargs[\"norm\"].vmin, 1, imshow_kwargs[\"norm\"].vmax], fontsize=6)\n",
    "cbar_ax.minorticks_off()\n",
    "cbar_ax.tick_params(length=1.5, pad=1)#,direction='out', length=6, width=2, colors='r', grid_color='r', grid_alpha=0.5)\n",
    "for _tidx, tick in enumerate(cbar_ax.xaxis.get_majorticklabels()):\n",
    "    if _tidx == 0:\n",
    "        tick.set_horizontalalignment(\"left\")\n",
    "    elif _tidx == 2:\n",
    "        tick.set_horizontalalignment(\"right\")\n",
    "    else:\n",
    "        tick.set_horizontalalignment(\"center\")\n",
    "\n",
    "_fig_fname = \"FigExt7I_mp.svg\"\n",
    "fig.savefig(_fig_fname, dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Older stuff that isn't publication ready ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # # # the mix one - mp\n",
    "sub_samples_m = [\n",
    "    \"N93m5\",\n",
    "    \"N93m10\",\n",
    "]\n",
    "# p ...\n",
    "sub_samples_p = [\n",
    "    \"N93p5\",\n",
    "    \"N93p10\",\n",
    "]\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=len(sub_samples_m),\n",
    "    ncols=2*len(distances),\n",
    "    figsize=(4*len(distances),2*len(sub_samples_m)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "\n",
    "imshow_kwargs = dict(\n",
    "        norm=LogNorm(vmin=1/2.5, vmax=2.5),\n",
    "        cmap=\"RdBu_r\",\n",
    "        interpolation=\"none\",\n",
    ")\n",
    "\n",
    "for sample_m, sample_p, (i, axs) in zip(sub_samples_m, sub_samples_p, enumerate(axs)):\n",
    "    for jj, (_dist_name, _dist) in enumerate(distances.items()):\n",
    "        axm, axp = axs[jj], axs[len(distances) + jj]\n",
    "        if _dist_name != \"trans\":\n",
    "            Cm = np.nanmean(interaction_sums[sample_m][_dist], axis=0) / np.nanmean(interaction_counts[sample_m][_dist], axis=0)\n",
    "            Cp = np.nanmean(interaction_sums[sample_p][_dist], axis=0) / np.nanmean(interaction_counts[sample_p][_dist], axis=0)\n",
    "        elif _dist_name == \"trans\":\n",
    "            Cm = np.nanmean(interaction_sums_trans[sample_m][_dist], axis=0) / np.nanmean(interaction_counts_trans[sample_m][_dist], axis=0)\n",
    "            Cp = np.nanmean(interaction_sums_trans[sample_p][_dist], axis=0) / np.nanmean(interaction_counts_trans[sample_p][_dist], axis=0)\n",
    "        else:\n",
    "            pass\n",
    "        axm.imshow(Cm, **imshow_kwargs)\n",
    "        axp.imshow(Cp, **imshow_kwargs)\n",
    "        for _ax in [axp, axm]:\n",
    "            _ax.set_xticks([])\n",
    "            _ax.set_yticks([])\n",
    "        if i == 0:\n",
    "            axm.set_title(f\"m-{_dist_name}\")\n",
    "            axp.set_title(f\"p-{_dist_name}\")\n",
    "        if i == len(sub_samples_m)-1:\n",
    "            for _ax in [axm, axp]:\n",
    "                _ax.set_xticks(np.arange(len(ticklabels)))\n",
    "                _ax.set_xticklabels(np.asarray(ticklabels[::-1]), rotation=\"vertical\")\n",
    "        if jj == 0:\n",
    "            axm.set_ylabel(sample_m.lstrip(\"m\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_samples_m =[\n",
    "        \"m10hR1R2\",\n",
    "        \"p10hR1R2\",\n",
    "        \"mp10hR1R2\",\n",
    "    ]\n",
    "sub_samples_p = [\n",
    "        \"N93m10\",\n",
    "        \"N93p10\",\n",
    "        \"N93mp10\",\n",
    "    ]\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=len(sub_samples_m),\n",
    "    ncols=2*len(distances),\n",
    "    figsize=(4*len(distances),2*len(sub_samples_m)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "\n",
    "imshow_kwargs = dict(\n",
    "        norm=LogNorm(vmin=1/2.5, vmax=2.5),\n",
    "        cmap=\"RdBu_r\",\n",
    "        interpolation=\"none\",\n",
    ")\n",
    "\n",
    "for sample_m, sample_p, (i, axs) in zip(sub_samples_m, sub_samples_p, enumerate(axs)):\n",
    "    for jj, (_dist_name, _dist) in enumerate(distances.items()):\n",
    "        axm, axp = axs[jj], axs[len(distances) + jj]\n",
    "        if _dist_name != \"trans\":\n",
    "            Cm = np.nanmean(interaction_sums[sample_m][_dist], axis=0) / np.nanmean(interaction_counts[sample_m][_dist], axis=0)\n",
    "            Cp = np.nanmean(interaction_sums[sample_p][_dist], axis=0) / np.nanmean(interaction_counts[sample_p][_dist], axis=0)\n",
    "        elif _dist_name == \"trans\":\n",
    "            Cm = np.nanmean(interaction_sums_trans[sample_m][_dist], axis=0) / np.nanmean(interaction_counts_trans[sample_m][_dist], axis=0)\n",
    "            Cp = np.nanmean(interaction_sums_trans[sample_p][_dist], axis=0) / np.nanmean(interaction_counts_trans[sample_p][_dist], axis=0)\n",
    "        else:\n",
    "            pass\n",
    "        axm.imshow(Cm, **imshow_kwargs)\n",
    "        axp.imshow(Cp, **imshow_kwargs)\n",
    "        for _ax in [axp, axm]:\n",
    "            _ax.set_xticks([])\n",
    "            _ax.set_yticks([])\n",
    "        if i == 0:\n",
    "            axm.set_title(f\"m-{_dist_name}\")\n",
    "            axp.set_title(f\"p-{_dist_name}\")\n",
    "        if i == len(sub_samples_m)-1:\n",
    "            for _ax in [axm, axp]:\n",
    "                _ax.set_xticks(np.arange(len(ticklabels)))\n",
    "                _ax.set_xticklabels(np.asarray(ticklabels[::-1]), rotation=\"vertical\")\n",
    "        if jj == 0:\n",
    "            axm.set_ylabel(sample_m)\n",
    "\n",
    "\n",
    "\n",
    "# try adding an axes manually ...\n",
    "cax = fig.add_axes([0.88,0.001,0.1,0.02])\n",
    "fig.colorbar(\n",
    "    cm.ScalarMappable(norm=imshow_kwargs[\"norm\"], cmap=imshow_kwargs[\"cmap\"]),\n",
    "    cax=cax,\n",
    "    orientation=\"horizontal\",\n",
    ")\n",
    "cax.set_xticks([imshow_kwargs[\"norm\"].vmin, 1, imshow_kwargs[\"norm\"].vmax])\n",
    "cax.set_xticklabels([imshow_kwargs[\"norm\"].vmin, 1, imshow_kwargs[\"norm\"].vmax])\n",
    "cax.minorticks_off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
