{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e7bd92b-9ad7-4627-8584-a627528ad3ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! mkdir genes_allana\n",
    "# ! scp newhpc:/home/allana.schooley-umw/as38w/Ranger/RNAseq/geneannotation/hg38.refGene.genpred.f.sorted.bed ./genes_allana/\n",
    "# ! scp newhpc:/home/allana.schooley-umw/as38w/Ranger/RNAseq/geneannotation/hg38.refGene.genpred.r.sorted.bed ./genes_allana/\n",
    "# ! scp newhpc:/home/allana.schooley-umw/as38w/Ranger/RNAseq/geneannotation/hg38.refGene.genpred.atRNAseq.fpkmg1.f.sorted.bed ./genes_allana/\n",
    "# ! scp newhpc:/home/allana.schooley-umw/as38w/Ranger/RNAseq/geneannotation/hg38.refGene.genpred.atRNAseq.fpkmg1.r.sorted.bed ./genes_allana/\n",
    "# ! head genes_allana/hg38.refGene.genpred.f.sorted.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13c34e20-d25c-4b86-a75e-18e5f062f2c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from BCBio.GFF import GFFExaminer\n",
    "# from BCBio import GFF\n",
    "# import pprint\n",
    "# from dna_features_viewer import BiopythonTranslator\n",
    "# ! head /data/sergpolly/hg38.gff\n",
    "# in_file = \"/data/sergpolly/hg38.gff\"\n",
    "# # examiner = GFFExaminer()\n",
    "# # in_handle = open(in_file)\n",
    "# # pprint.pprint(examiner.parent_child_map(in_handle))\n",
    "# # in_handle.close\n",
    "# examiner = GFFExaminer()\n",
    "# in_handle = open(in_file)\n",
    "# pprint.pprint(examiner.available_limits(in_handle))\n",
    "# in_handle.close()\n",
    "\n",
    "\n",
    "# in_file = \"/data/sergpolly/hg38.gff\"\n",
    "# # examiner = GFFExaminer()\n",
    "# # in_handle = open(in_file)\n",
    "# # pprint.pprint(examiner.available_limits(in_handle))\n",
    "# # in_handle.close()\n",
    "\n",
    "# _gff_filter = dict(\n",
    "#     gff_id=['NC_000006.12'],  # chr6 - maybe ?!\n",
    "#     # gff_source=['RefSeq'],\n",
    "#     gff_type=[\"gene\"],\n",
    "# )\n",
    "\n",
    "# in_handle = open(in_file)\n",
    "# for rec in GFF.parse( in_handle, limit_info=_gff_filter ):\n",
    "#     _chr6genes = rec\n",
    "#     break\n",
    "# in_handle.close()\n",
    "\n",
    "# # # # ...\n",
    "# # region1_10 = ('chr6', 129100000-230_000, 130500000+30_000)\n",
    "# # region2_10 = ('chr6', 136100000-45_000, 138200000+45_000)\n",
    "\n",
    "\n",
    "# from dna_features_viewer import BiopythonTranslator\n",
    "# # record = SeqIO.read(\"example_sequence.gb\", \"genbank\")\n",
    "\n",
    "# fplus = lambda f: f.location.strand > 0\n",
    "# fminus = lambda f: f.location.strand < 0\n",
    "# fsize = lambda f: (f.location.end - f.location.start)>15_000\n",
    "\n",
    "# translator_plus = BiopythonTranslator(features_filters=[fplus, fsize])\n",
    "# translator_minus = BiopythonTranslator(features_filters=[fminus, fsize])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0df3572-49ef-4689-bb70-5d9dcb3eb60b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# _start, _end = 136100000, 138200000\n",
    "# _region = ('chr6', _start, _end)\n",
    "\n",
    "# gfs_fwd = []\n",
    "# for _row in bioframe.merge(bioframe.select(_fwd, _region)).itertuples(index=False):\n",
    "#     # print(_row)\n",
    "#     _gf = GraphicFeature(\n",
    "#         start=int(_row.start),\n",
    "#         end=int(_row.end),\n",
    "#         strand=1,\n",
    "#         color='black',\n",
    "#         **gfargs,\n",
    "#     )\n",
    "#     gfs_fwd.append(_gf)\n",
    "\n",
    "\n",
    "# gfs_rev = []\n",
    "# for _row in bioframe.merge(bioframe.select(_rev, _region)).itertuples(index=False):\n",
    "#     # print(_row)\n",
    "#     _gf = GraphicFeature(\n",
    "#         start=int(_row.start),\n",
    "#         end=int(_row.end),\n",
    "#         strand=-1,\n",
    "#         color='black',\n",
    "#         **gfargs\n",
    "#    )\n",
    "#     gfs_rev.append(_gf)\n",
    "\n",
    "\n",
    "\n",
    "# gfs_act_fwd = []\n",
    "# for _row in bioframe.merge(bioframe.select(_fwd_act, _region)).itertuples(index=False):\n",
    "#     # print(_row)\n",
    "#     _gf = GraphicFeature(\n",
    "#         start=int(_row.start),\n",
    "#         end=int(_row.end),\n",
    "#         strand=1,\n",
    "#         color='green',\n",
    "#         **gfargs,\n",
    "#     )\n",
    "#     gfs_act_fwd.append(_gf)\n",
    "\n",
    "\n",
    "# gfs_act_rev = []\n",
    "# for _row in bioframe.merge(bioframe.select(_rev_act, _region)).itertuples(index=False):\n",
    "#     # print(_row)\n",
    "#     _gf = GraphicFeature(\n",
    "#         start=int(_row.start),\n",
    "#         end=int(_row.end),\n",
    "#         strand=-1,\n",
    "#         color='green',\n",
    "#         **gfargs\n",
    "#    )\n",
    "#     gfs_act_rev.append(_gf)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# _hui_fwd = GraphicRecord(\n",
    "#     sequence_length=(_end-_start+1),\n",
    "#     features=gfs_fwd,\n",
    "# )\n",
    "\n",
    "# _hui_rev = GraphicRecord(\n",
    "#     sequence_length=(_end-_start+1),\n",
    "#     features=gfs_rev,\n",
    "# )\n",
    "\n",
    "\n",
    "# _hui_act_fwd = GraphicRecord(\n",
    "#     sequence_length=(_end-_start+1),\n",
    "#     features=gfs_act_fwd,\n",
    "# )\n",
    "\n",
    "# _hui_act_rev = GraphicRecord(\n",
    "#     sequence_length=(_end-_start+1),\n",
    "#     features=gfs_act_rev,\n",
    "# )\n",
    "\n",
    "# # f,ax = plt.subplots(1,1)\n",
    "# ax, _ = _hui_fwd.plot(\n",
    "#     x_lim=(_start, _end),\n",
    "#     level_offset=-0.5,\n",
    "#     **plotargs,\n",
    "# )\n",
    "\n",
    "# _hui_rev.plot(\n",
    "#     ax=ax,\n",
    "#     x_lim=(_start, _end),\n",
    "#     level_offset=0.5,\n",
    "#     **plotargs,\n",
    "# )\n",
    "\n",
    "\n",
    "# _hui_act_fwd.plot(\n",
    "#     ax=ax,\n",
    "#     x_lim=(_start, _end),\n",
    "#     level_offset=-0.5,\n",
    "#     **plotargs,\n",
    "# )\n",
    "# _hui_act_rev.plot(\n",
    "#     ax=ax,\n",
    "#     x_lim=(_start, _end),\n",
    "#     level_offset=0.5,\n",
    "#     **plotargs,\n",
    "# )\n",
    "# ax.set_ylim(-1,1)\n",
    "# ax.set_axis_on()\n",
    "# # ax.set_yticks([-1,0,1,2,3,4,5])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e3e7973",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import chain\n",
    "\n",
    "# Hi-C utilities imports:\n",
    "import cooler\n",
    "import bioframe\n",
    "import cooltools\n",
    "from cooltools.lib.numutils import fill_diag\n",
    "\n",
    "# Visualization imports:\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LogNorm\n",
    "from matplotlib.colors import PowerNorm\n",
    "from matplotlib import cm\n",
    "# %matplotlib widget\n",
    "%matplotlib inline\n",
    "\n",
    "import matplotlib.patches as patches\n",
    "from matplotlib.ticker import EngFormatter\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from cooltools.lib import plotting\n",
    "\n",
    "from itertools import cycle\n",
    "\n",
    "# # from ipywidgets import interact, fixed\n",
    "# from ipywidgets import interact, interactive, fixed, interact_manual\n",
    "# import ipywidgets as widgets\n",
    "\n",
    "# enable editable text ...\n",
    "mpl.rcParams[\"pdf.fonttype\"]=42\n",
    "mpl.rcParams[\"svg.fonttype\"]='none'\n",
    "mpl.rcParams['axes.linewidth'] = 0.5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7d743c5",
   "metadata": {},
   "source": [
    "### Import modified \"guts\" of the dotfinder submodule from `helper_func` file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e1c6ae5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from helper_func import draw_kernel\n",
    "\n",
    "# turns out still need some of the dotfinder guts in here\n",
    "from cooltools.api.dotfinder import bp_to_bins, generate_tiles_diag_band\n",
    "from cooltools.lib.numutils import LazyToeplitz\n",
    "from cooltools.lib.common import assign_regions\n",
    "\n",
    "from functools import partial\n",
    "from data_catalog import bws, bws_vlim, telo_dict\n",
    "\n",
    "import datashader as ds, xarray as xr\n",
    "from datashader import transfer_functions as tf, reductions as rd\n",
    "from datashader.mpl_ext import dsshow, alpha_colormap\n",
    "\n",
    "from matplotlib.cm import YlOrRd\n",
    "import matplotlib.image as mpimg\n",
    "\n",
    "from palettable.scientific import sequential\n",
    "lajolla_r = sequential.LaJolla_10_r.get_mpl_colormap()\n",
    "bilbao_r = sequential.Bilbao_20_r.get_mpl_colormap()\n",
    "batlow_r = sequential.Batlow_20_r.get_mpl_colormap()\n",
    "devon_r = sequential.Devon_20_r.get_mpl_colormap()\n",
    "davos_r = sequential.Davos_20_r.get_mpl_colormap()\n",
    "# fall_cmap = fall\n",
    "lajolla = sequential.LaJolla_10.get_mpl_colormap()\n",
    "bilbao = sequential.Bilbao_20.get_mpl_colormap()\n",
    "batlow = sequential.Batlow_20.get_mpl_colormap()\n",
    "devon = sequential.Devon_20.get_mpl_colormap()\n",
    "davos = sequential.Davos_20.get_mpl_colormap()\n",
    "\n",
    "imola_r = sequential.Imola_20_r.get_mpl_colormap()\n",
    "imola = sequential.Imola_20.get_mpl_colormap()\n",
    "\n",
    "nuuk_r = sequential.Nuuk_20_r.get_mpl_colormap()\n",
    "nuuk = sequential.Nuuk_20.get_mpl_colormap()\n",
    "\n",
    "hawaii_r = sequential.Hawaii_20_r.get_mpl_colormap()\n",
    "hawaii = sequential.Hawaii_20.get_mpl_colormap()\n",
    "\n",
    "oleron_r = sequential.Oleron_20_r.get_mpl_colormap()\n",
    "oleron = sequential.Oleron_20.get_mpl_colormap()\n",
    "\n",
    "oslo_r = sequential.Oslo_20_r.get_mpl_colormap()\n",
    "oslo = sequential.Oslo_20.get_mpl_colormap()\n",
    "\n",
    "bamako_r = sequential.Bamako_20_r.get_mpl_colormap()\n",
    "bamako = sequential.Bamako_20.get_mpl_colormap()\n",
    "\n",
    "grayc_r = sequential.GrayC_20_r.get_mpl_colormap()\n",
    "grayc = sequential.GrayC_20.get_mpl_colormap()\n",
    "\n",
    "lapaz_r = sequential.LaPaz_20_r.get_mpl_colormap()\n",
    "lapaz = sequential.LaPaz_20.get_mpl_colormap()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f941017-971c-4bb6-8cc9-3cdbbeac3703",
   "metadata": {},
   "outputs": [],
   "source": [
    "! ls genes_allana"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "881b6864-c7aa-4dc6-829f-d54c803b9def",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dna_features_viewer import GraphicRecord, GraphicFeature\n",
    "fwd_args = dict(\n",
    "    label=None,\n",
    "    thickness=4,\n",
    "    linewidth=0,\n",
    "    strand=1,\n",
    "    color=\"tab:red\",\n",
    ")\n",
    "rev_args = dict(\n",
    "    label=None,\n",
    "    thickness=4,\n",
    "    linewidth=0,\n",
    "    strand=-1,\n",
    "    color=\"tab:blue\",\n",
    ")\n",
    "\n",
    "plotargs = dict(\n",
    "    with_ruler=False,\n",
    "    draw_line=False,\n",
    ")\n",
    "\n",
    "_fwd = bioframe.read_table(\"genes_allana/hg38.refGene.genpred.f.sorted.bed\",schema=\"bed12\").eval(\"size = end - start\").query(\"size > 8_000\")\n",
    "_rev = bioframe.read_table(\"genes_allana/hg38.refGene.genpred.r.sorted.bed\",schema=\"bed12\").eval(\"size = end - start\").query(\"size > 8_000\")\n",
    "\n",
    "_fwd_act = bioframe.read_table(\"genes_allana/hg38.refGene.genpred.atRNAseq.fpkmg1.f.sorted.bed\",schema=\"bed12\").eval(\"size = end - start\").query(\"size > 8_000\")\n",
    "_rev_act = bioframe.read_table(\"genes_allana/hg38.refGene.genpred.atRNAseq.fpkmg1.r.sorted.bed\",schema=\"bed12\").eval(\"size = end - start\").query(\"size > 8_000\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fee642b6-6d45-41fb-8094-49ad2798c7e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "_fwd_inact = bioframe.setdiff(\n",
    "    _fwd,\n",
    "    _fwd_act\n",
    ")\n",
    "\n",
    "_rev_inact = bioframe.setdiff(\n",
    "    _rev,\n",
    "    _rev_act\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1c5b215-e96b-4409-a158-7547ecec6460",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy import interpolate\n",
    "# https://stackoverflow.com/questions/6518811/interpolate-nan-values-in-a-numpy-array\n",
    "def fill_nan(A):\n",
    "    '''\n",
    "    interpolate to fill nan values\n",
    "    '''\n",
    "    inds = np.arange(A.shape[0])\n",
    "    good = np.where(np.isfinite(A))\n",
    "    f = interpolate.interp1d(inds[good], A[good],bounds_error=False)\n",
    "    B = np.where(np.isfinite(A),A,f(inds))\n",
    "    return B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce5554c6-8a81-443f-aeb9-a2fb6877dfc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import matplotlib.pyplot as plt\n",
    "import bbi\n",
    "import matplotlib.lines as lines\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.patches import ConnectionPatch, Rectangle\n",
    "from mpl_toolkits.axes_grid1 import Divider, Size\n",
    "from mpl_toolkits.axes_grid1.inset_locator import BboxConnector\n",
    "\n",
    "# manual colorbar example ...\n",
    "# #\n",
    "# cax = divider.append_axes(\"right\", size=0.1, pad=0.05)\n",
    "# plt.colorbar(\n",
    "#     cm.ScalarMappable(norm=PowerNorm(1/3., vmin=vmin, vmax=vmax), cmap=mycmap),\n",
    "#     cax=cax,\n",
    "# )\n",
    "\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.path as mpath\n",
    "\n",
    "# import matplotlib.pyplot as plt\n",
    "import bbi\n",
    "import matplotlib.lines as lines\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.patches import ConnectionPatch, Rectangle\n",
    "from mpl_toolkits.axes_grid1 import Divider, Size\n",
    "from mpl_toolkits.axes_grid1.inset_locator import BboxConnector\n",
    "\n",
    "\n",
    "# draw 2 arm chromosome ....\n",
    "def get_chrom_path(\n",
    "    x0y0,  # position of the centromere ...\n",
    "    parm,  # p arm length\n",
    "    qarm,  # q arm length\n",
    "    width,  # width\n",
    "    color=\"grey\",\n",
    "):\n",
    "    # unpack coordinates of the center (centromere)\n",
    "    x0, y0 = x0y0\n",
    "\n",
    "    # calculate half width\n",
    "    hw = 0.5*width\n",
    "\n",
    "    dtelo = 0.5*width\n",
    "    dxcent = 0.2*width\n",
    "    dycent = 0.5*width\n",
    "\n",
    "    # Prepare the data for the PathPatch below.\n",
    "    Path = mpath.Path\n",
    "    codes, verts = zip(*[\n",
    "        # p-arm up:\n",
    "        (Path.MOVETO, [x0-hw, y0+dycent]),\n",
    "        (Path.LINETO, [x0-hw, y0+parm-dtelo]),\n",
    "        # p-telomere curve:\n",
    "        (Path.CURVE3, [x0-hw, y0+parm]),\n",
    "        (Path.CURVE3, [x0, y0+parm]),\n",
    "        (Path.CURVE3, [x0+hw, y0+parm]),\n",
    "        (Path.CURVE3, [x0+hw, y0+parm-dtelo]),\n",
    "        # p-arm down:\n",
    "        (Path.LINETO, [x0+hw, y0+dycent]),\n",
    "        # centromere S curve (right)\n",
    "        # (1S)\n",
    "        (Path.CURVE4, [x0+hw, y0+0.5*dycent]),\n",
    "        (Path.CURVE4, [x0+hw-dxcent, y0+0.5*dycent]),\n",
    "        (Path.CURVE4, [x0+hw-dxcent, y0]),\n",
    "        # (2S)\n",
    "        (Path.CURVE4, [x0+hw-dxcent, y0-0.5*dycent]),\n",
    "        (Path.CURVE4, [x0+hw, y0-0.5*dycent]),\n",
    "        (Path.CURVE4, [x0+hw, y0-dycent]),\n",
    "        # q-arm down\n",
    "        (Path.LINETO, [x0+hw, y0-qarm+dtelo]),\n",
    "        # q-telomere curve:\n",
    "        (Path.CURVE3, [x0+hw, y0-qarm]),\n",
    "        (Path.CURVE3, [x0, y0-qarm]),\n",
    "        (Path.CURVE3, [x0-hw, y0-qarm]),\n",
    "        (Path.CURVE3, [x0-hw, y0-qarm+dtelo]),\n",
    "        # q-arm up\n",
    "        (Path.LINETO, [x0-hw, y0-dycent]),\n",
    "        # centromere S curve (left)\n",
    "        (Path.CURVE4, [x0-hw, y0-0.5*dycent]),\n",
    "        (Path.CURVE4, [x0-hw+dxcent, y0-0.5*dycent]),\n",
    "        (Path.CURVE4, [x0-hw+dxcent, y0]),\n",
    "        (Path.CURVE4, [x0-hw+dxcent, y0+0.5*dycent]),\n",
    "        (Path.CURVE4, [x0-hw, y0+0.5*dycent]),\n",
    "        (Path.CURVE4, [x0-hw, y0+dycent]),\n",
    "        (Path.CLOSEPOLY, [x0-hw, y0+dycent]),\n",
    "    ])\n",
    "\n",
    "    chrom = mpatches.PathPatch(mpath.Path(verts, codes), ec=\"none\")\n",
    "    chrom.set(color=color)\n",
    "    return chrom\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# draw 2 arm chromosome ....\n",
    "def get_chrom_path_wpathes(\n",
    "    x0y0,  # position of the centromere ...\n",
    "    parm,  # p arm length\n",
    "    qarm,  # q arm length\n",
    "    width,  # width\n",
    "    color=\"grey\",\n",
    "    patches=None,  # [(_arm, _from, _to, pcolor)]\n",
    "):\n",
    "    # unpack coordinates of the center (centromere)\n",
    "    x0, y0 = x0y0\n",
    "\n",
    "    # calculate half width\n",
    "    hw = 0.5*width\n",
    "\n",
    "    dtelo = 0.5*width\n",
    "    dxcent = 0.2*width\n",
    "    dycent = 0.5*width\n",
    "\n",
    "    # Prepare the data for the PathPatch below.\n",
    "    Path = mpath.Path\n",
    "    codes, verts = zip(*[\n",
    "        # p-arm up:\n",
    "        (Path.MOVETO, [x0-hw, y0+dycent]),\n",
    "        (Path.LINETO, [x0-hw, y0+parm-dtelo]),\n",
    "        # p-telomere curve:\n",
    "        (Path.CURVE3, [x0-hw, y0+parm]),\n",
    "        (Path.CURVE3, [x0, y0+parm]),\n",
    "        (Path.CURVE3, [x0+hw, y0+parm]),\n",
    "        (Path.CURVE3, [x0+hw, y0+parm-dtelo]),\n",
    "        # p-arm down:\n",
    "        (Path.LINETO, [x0+hw, y0+dycent]),\n",
    "        # centromere S curve (right)\n",
    "        # (1S)\n",
    "        (Path.CURVE4, [x0+hw, y0+0.5*dycent]),\n",
    "        (Path.CURVE4, [x0+hw-dxcent, y0+0.5*dycent]),\n",
    "        (Path.CURVE4, [x0+hw-dxcent, y0]),\n",
    "        # (2S)\n",
    "        (Path.CURVE4, [x0+hw-dxcent, y0-0.5*dycent]),\n",
    "        (Path.CURVE4, [x0+hw, y0-0.5*dycent]),\n",
    "        (Path.CURVE4, [x0+hw, y0-dycent]),\n",
    "        # q-arm down\n",
    "        (Path.LINETO, [x0+hw, y0-qarm+dtelo]),\n",
    "        # q-telomere curve:\n",
    "        (Path.CURVE3, [x0+hw, y0-qarm]),\n",
    "        (Path.CURVE3, [x0, y0-qarm]),\n",
    "        (Path.CURVE3, [x0-hw, y0-qarm]),\n",
    "        (Path.CURVE3, [x0-hw, y0-qarm+dtelo]),\n",
    "        # q-arm up\n",
    "        (Path.LINETO, [x0-hw, y0-dycent]),\n",
    "        # centromere S curve (left)\n",
    "        (Path.CURVE4, [x0-hw, y0-0.5*dycent]),\n",
    "        (Path.CURVE4, [x0-hw+dxcent, y0-0.5*dycent]),\n",
    "        (Path.CURVE4, [x0-hw+dxcent, y0]),\n",
    "        (Path.CURVE4, [x0-hw+dxcent, y0+0.5*dycent]),\n",
    "        (Path.CURVE4, [x0-hw, y0+0.5*dycent]),\n",
    "        (Path.CURVE4, [x0-hw, y0+dycent]),\n",
    "        (Path.CLOSEPOLY, [x0-hw, y0+dycent]),\n",
    "    ])\n",
    "\n",
    "    chrom = mpatches.PathPatch(mpath.Path(verts, codes), ec=\"none\")\n",
    "    chrom.set(color=color)\n",
    "\n",
    "    patches_to_draw = []\n",
    "    if patches is not None:\n",
    "        # go over patches ...\n",
    "        for _arm, _from, _to, pcolor in patches:\n",
    "            hw = hw + 0.005\n",
    "            if _arm == \"q\":\n",
    "                codes, verts = zip(*[\n",
    "                    # rectangle\n",
    "                    (Path.MOVETO, [x0-hw, y0-_from]),\n",
    "                    (Path.LINETO, [x0-hw, y0-_to]),\n",
    "                    (Path.LINETO, [x0+hw, y0-_to]),\n",
    "                    (Path.LINETO, [x0+hw, y0-_from]),\n",
    "                    (Path.CLOSEPOLY, [x0-hw, y0-_from]),\n",
    "                ])\n",
    "            elif _arm == \"p\":\n",
    "                codes, verts = zip(*[\n",
    "                    # rectangle\n",
    "                    (Path.MOVETO, [x0-hw, y0+_from]),\n",
    "                    (Path.LINETO, [x0-hw, y0+_to]),\n",
    "                    (Path.LINETO, [x0+hw, y0+_to]),\n",
    "                    (Path.LINETO, [x0+hw, y0+_from]),\n",
    "                    (Path.CLOSEPOLY, [x0-hw, y0+_from]),\n",
    "                ])\n",
    "            else:\n",
    "                raise(\"arms !!!\")\n",
    "            part = mpatches.PathPatch(mpath.Path(verts, codes), ec=\"none\")\n",
    "            part.set(color=pcolor)\n",
    "            patches_to_draw.append(part)\n",
    "\n",
    "    return chrom, patches_to_draw\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# draw 2 arm chromosome ....\n",
    "def get_chrom_path_highlight(\n",
    "    x0y0,  # position of the centromere ...\n",
    "    parm,  # p arm length\n",
    "    qarm,  # q arm length\n",
    "    width,  # width\n",
    "    color=\"grey\",\n",
    "    pshade=0,  # fraction of p arm to be highlighted\n",
    "    qshade=0,  # fraction of q arm to be highlighted\n",
    "    colorp=\"black\",\n",
    "    colorq=\"black\",\n",
    "):\n",
    "    \"\"\"\n",
    "    a function that also highlight terminal part of\n",
    "    p and q arms ...\n",
    "    \"\"\"\n",
    "    # unpack coordinates of the center (centromere)\n",
    "    x0, y0 = x0y0\n",
    "\n",
    "    # calculate half width\n",
    "    hw = 0.5*width\n",
    "\n",
    "    dtelo = 0.5*width\n",
    "    dxcent = 0.2*width\n",
    "    dycent = 0.5*width\n",
    "\n",
    "    # Prepare the data for the PathPatch below.\n",
    "    Path = mpath.Path\n",
    "    codes, verts = zip(*[\n",
    "        # p-arm up:\n",
    "        (Path.MOVETO, [x0-hw, y0+dycent]),\n",
    "        (Path.LINETO, [x0-hw, y0+parm-dtelo]),\n",
    "        # p-telomere curve:\n",
    "        (Path.CURVE3, [x0-hw, y0+parm]),\n",
    "        (Path.CURVE3, [x0, y0+parm]),\n",
    "        (Path.CURVE3, [x0+hw, y0+parm]),\n",
    "        (Path.CURVE3, [x0+hw, y0+parm-dtelo]),\n",
    "        # p-arm down:\n",
    "        (Path.LINETO, [x0+hw, y0+dycent]),\n",
    "        # centromere S curve (right)\n",
    "        # (1S)\n",
    "        (Path.CURVE4, [x0+hw, y0+0.5*dycent]),\n",
    "        (Path.CURVE4, [x0+hw-dxcent, y0+0.5*dycent]),\n",
    "        (Path.CURVE4, [x0+hw-dxcent, y0]),\n",
    "        # (2S)\n",
    "        (Path.CURVE4, [x0+hw-dxcent, y0-0.5*dycent]),\n",
    "        (Path.CURVE4, [x0+hw, y0-0.5*dycent]),\n",
    "        (Path.CURVE4, [x0+hw, y0-dycent]),\n",
    "        # q-arm down\n",
    "        (Path.LINETO, [x0+hw, y0-qarm+dtelo]),\n",
    "        # q-telomere curve:\n",
    "        (Path.CURVE3, [x0+hw, y0-qarm]),\n",
    "        (Path.CURVE3, [x0, y0-qarm]),\n",
    "        (Path.CURVE3, [x0-hw, y0-qarm]),\n",
    "        (Path.CURVE3, [x0-hw, y0-qarm+dtelo]),\n",
    "        # q-arm up\n",
    "        (Path.LINETO, [x0-hw, y0-dycent]),\n",
    "        # centromere S curve (left)\n",
    "        (Path.CURVE4, [x0-hw, y0-0.5*dycent]),\n",
    "        (Path.CURVE4, [x0-hw+dxcent, y0-0.5*dycent]),\n",
    "        (Path.CURVE4, [x0-hw+dxcent, y0]),\n",
    "        (Path.CURVE4, [x0-hw+dxcent, y0+0.5*dycent]),\n",
    "        (Path.CURVE4, [x0-hw, y0+0.5*dycent]),\n",
    "        (Path.CURVE4, [x0-hw, y0+dycent]),\n",
    "        (Path.CLOSEPOLY, [x0-hw, y0+dycent]),\n",
    "    ])\n",
    "    chrom = mpatches.PathPatch(mpath.Path(verts, codes), ec=\"none\")\n",
    "    chrom.set(color=color)\n",
    "\n",
    "    hw = hw + 0.005\n",
    "    if pshade > 0:\n",
    "        dy_pshade = (1.0-pshade)*parm\n",
    "        if dy_pshade < dycent:\n",
    "            dy_pshade = dycent\n",
    "        # p -arm highlight\n",
    "        codes, verts = zip(*[\n",
    "            # p-arm up:\n",
    "            (Path.MOVETO, [x0-hw, y0+dy_pshade]),\n",
    "            (Path.LINETO, [x0-hw, y0+parm-dtelo]),\n",
    "            # p-telomere curve:\n",
    "            (Path.CURVE3, [x0-hw, y0+parm]),\n",
    "            (Path.CURVE3, [x0, y0+parm]),\n",
    "            (Path.CURVE3, [x0+hw, y0+parm]),\n",
    "            (Path.CURVE3, [x0+hw, y0+parm-dtelo]),\n",
    "            # p-arm down:\n",
    "            (Path.LINETO, [x0+hw, y0+dy_pshade]),\n",
    "            (Path.CLOSEPOLY, [x0-hw, y0+dy_pshade]),\n",
    "        ])\n",
    "        part = mpatches.PathPatch(mpath.Path(verts, codes), ec=\"none\")\n",
    "        part.set(color=colorp)\n",
    "    else:\n",
    "        part = None\n",
    "        dy_pshade = 0\n",
    "\n",
    "    if qshade > 0:\n",
    "        # Prepare the data for the PathPatch below.\n",
    "        dy_qshade = (1.0-qshade)*qarm\n",
    "        if dy_qshade < dycent:\n",
    "            dy_qshade = dycent\n",
    "        codes, verts = zip(*[\n",
    "            (Path.MOVETO, [x0+hw, y0-dy_qshade]),\n",
    "            # q-arm down\n",
    "            (Path.LINETO, [x0+hw, y0-qarm+dtelo]),\n",
    "            # q-telomere curve:\n",
    "            (Path.CURVE3, [x0+hw, y0-qarm]),\n",
    "            (Path.CURVE3, [x0, y0-qarm]),\n",
    "            (Path.CURVE3, [x0-hw, y0-qarm]),\n",
    "            (Path.CURVE3, [x0-hw, y0-qarm+dtelo]),\n",
    "            # q-arm up\n",
    "            (Path.LINETO, [x0-hw, y0-dy_qshade]),\n",
    "            (Path.CLOSEPOLY, [x0+hw, y0-dy_qshade]),\n",
    "        ])\n",
    "        qart = mpatches.PathPatch(mpath.Path(verts, codes), ec=\"none\")\n",
    "        qart.set(color=colorq)\n",
    "    else:\n",
    "        qart = None\n",
    "        dy_qshade = 0\n",
    "\n",
    "    return chrom, (part, y0+dy_pshade), (qart, y0-dy_qshade)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6727d9cd",
   "metadata": {},
   "source": [
    "### pick a dataset and binsize to work on ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "474819ac",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ! ls /data/proj_sync/data_ranger/finalcoolers_mega/ranGAP1-aux-G1s-MEGA.hg38.mapq_30.1000.mcool\n",
    "\n",
    "# 10 kb is a resolution at which one can clearly see \"dots\":\n",
    "binsize10 = 10_000\n",
    "# cooler files that we'll work on :\n",
    "telo_clrs10 = { _k: cooler.Cooler(f\"{_path}::/resolutions/{binsize10}\") for _k, _path in telo_dict.items() }\n",
    "\n",
    "# 25 and 50 ....\n",
    "binsize25 = 25_000\n",
    "# cooler files that we'll work on :\n",
    "telo_clrs25 = { _k: cooler.Cooler(f\"{_path}::/resolutions/{binsize25}\") for _k, _path in telo_dict.items() }\n",
    "\n",
    "binsize50 = 50_000\n",
    "# cooler files that we'll work on :\n",
    "telo_clrs50 = { _k: cooler.Cooler(f\"{_path}::/resolutions/{binsize50}\") for _k, _path in telo_dict.items() }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69f25e50-0477-4e03-bfd5-2e96616cf59d",
   "metadata": {},
   "source": [
    "## Define chromosome arms just in case ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d42f507-ff09-4571-a78e-ad01a4dafa86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define genomic view that will be used to call dots and pre-compute expected\n",
    "\n",
    "# Use bioframe to fetch the genomic features from the UCSC.\n",
    "hg38_chromsizes = bioframe.fetch_chromsizes('hg38')\n",
    "hg38_cens = bioframe.fetch_centromeres('hg38')\n",
    "hg38_arms_full = bioframe.make_chromarms(hg38_chromsizes, hg38_cens)\n",
    "# # remove \"bad\" chromosomes and near-empty arms ...\n",
    "included_arms = hg38_arms_full[\"name\"].to_list()[:44] # all autosomal ones ...\n",
    "hg38_arms = hg38_arms_full[hg38_arms_full[\"name\"].isin(included_arms)].reset_index(drop=True)\n",
    "arms_dict = hg38_arms_full.set_index(\"name\").eval(\"(end - start)//1000000\").to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e404bbe7-2f1a-453e-b079-8f61445c8297",
   "metadata": {},
   "source": [
    "### generate some bintables for cres and such ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "686de2b0-27f0-4c30-b90b-d01894715b72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate some binsized bins\n",
    "clr_bins10 = cooler.binnify(hg38_chromsizes, binsize10)\n",
    "clr_bins10 = clr_bins10[~clr_bins10[\"chrom\"].isin([\"chrX\",\"chrY\",\"chrM\"])]\n",
    "# now annotate those bins with useful info - e.g. ID anchors ...\n",
    "display(clr_bins10.head(2))\n",
    "display(clr_bins10.tail(2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6426a54-2bd2-4d43-85f4-f53ebc7298af",
   "metadata": {},
   "source": [
    "## Figure out anchors drawing and highlighting here ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ecd159c-02b6-4a8d-b586-76b24379c96f",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_anchor_fnames = {\n",
    "    \"mega_2X_enrichment\": \"ID_anchors/mega_2X_enrichment.fourth_mega.max_size.bed\",\n",
    "    \"5hr_2X_enrichment_old\": \"ID_anchors/5hr_2X_enrichment.second_bulk.max_size.bed\",\n",
    "    \"5hr_2X_enrichment\": \"ID_anchors/5hr_2X_enrichment.pixel_derived.bed\",\n",
    "    \"5hr_2X_enrichment_nosing\": \"ID_anchors/5hr_2X_enrichment.pixel_derived.no_singletons.bed\",\n",
    "    \"5hr_2X_enrichment_signal\": \"ID_anchors/5hr_2X_enrichment.pixel_derived.signal_peaks.bed\",\n",
    "    \"10hr_2X_enrichment_signal\": \"ID_anchors/10hrs_2X_enrichment.pixel_derived.signal_peaks.bed\",\n",
    "    \"N93p5_2X_enrichment_signal\": \"ID_anchors/N93p5_2X_enrichment.pixel_derived.signal_peaks.bed\",\n",
    "    \"pCyto_2X_enrichment_signal\": \"ID_anchors/pCyto_2X_enrichment.pixel_derived.signal_peaks.bed\",\n",
    "    \"mCyto_2X_enrichment_signal\": \"ID_anchors/mCyto_2X_enrichment.pixel_derived.signal_peaks.bed\",\n",
    "    \"mega_3X_enrichment\": \"ID_anchors/mega_3X_enrichment.fifth_mega3x.max_size.bed\",\n",
    "    \"MEGA_2X_enrichment\": \"ID_anchors/MEGAp5_2X_enrichment.pixel_derived.signal_peaks.bed\",\n",
    "    \"MEGAN93_2X_enrichment\": \"ID_anchors/MEGAN93p5_2X_enrichment.pixel_derived.signal_peaks.bed\",\n",
    "    \"MEGAminus_2X_enrichment\": \"ID_anchors/MEGA_minus_ctrl_2X_enrichment.pixel_derived.signal_peaks.bed\",\n",
    "    \"cyto_2x_enrichment\": \"ID_anchors/cyto_2x_enrichment.third_mCyto.max_size.bed\",\n",
    "}\n",
    "\n",
    "id_anchors_dict = {}\n",
    "for id_name, fname in id_anchor_fnames.items():\n",
    "    id_anchors_dict[id_name] = pd.read_csv(fname, sep=\"\\t\")\n",
    "    # ...\n",
    "    print(f\"loaded {len(id_anchors_dict[id_name]):5d} ID anchors {id_name:>20} in BED format ...\")\n",
    "\n",
    "_anchors = id_anchors_dict[\"5hr_2X_enrichment_signal\"]\n",
    "\n",
    "# we need bins/pixels for plotting - instead of genomic coordinates ...\n",
    "\n",
    "# need a reference cooler for that to do genomic coords -> bins:\n",
    "_clr = telo_clrs10[\"m5hR1R2\"]\n",
    "# doing that using clr.offset and clr.extent functionality:\n",
    "_anchors[\"bin_id\"] = _anchors[['chrom', 'peak_start', 'peak_end']].apply(_clr.offset, axis=1, result_type=\"expand\")\n",
    "_anchors[\"bin_width\"] = _anchors[['chrom', 'peak_start', 'peak_end']] \\\n",
    "    .apply(_clr.extent, axis=1, result_type=\"expand\") \\\n",
    "    .apply(np.diff, axis=1, result_type=\"expand\")[0]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "419fed08-2d96-45bb-a93a-0bb8e4710875",
   "metadata": {},
   "source": [
    "# Load domains\n",
    "merge nested and \"almost\" nested regions (regions overlapping a lot !)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce8d04e2-09a1-4fb8-a08b-2f75e6a5b6d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "_domains = pd.read_csv(\n",
    "    \"extrusion_domains/mG1s_MEGA_10kb_double_filtered.bedpe\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "display(_domains.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "082b1256-4fa4-4e10-b2fe-c94a0e4e7f2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mycmap = mpl.colors.LinearSegmentedColormap.from_list(\n",
    "    name=\"myfall\",\n",
    "    colors=[ mpl.colors.to_hex(c) for c in plotting.PALETTES[\"fall\"] ],\n",
    "    N=2**16,\n",
    ")\n",
    "mycmap.set_bad('white',.5)\n",
    "mycmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bd646e0-7c01-46aa-8173-e2c52efc5eb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "_idid = bioframe.pair_by_distance(\n",
    "    _anchors,\n",
    "    min_sep=0,\n",
    "    max_sep=200_000_000,  # should be enough ...\n",
    "    suffixes=(\"1\",\"2\"),\n",
    "    keep_order=True,\n",
    ")\n",
    "\n",
    "_idid = _idid.rename(columns={\"bin_id1\":\"bin1_id\", \"bin_id2\":\"bin2_id\", \"bin_width1\":\"bin1_width\", \"bin_width2\":\"bin2_width\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd15c333",
   "metadata": {},
   "source": [
    "# Working on a Figure here\n",
    "\n",
    "250kb (incl trans)\n",
    " - region1 = ('chr6', 105000000, 155000000)\n",
    " - region2 = ('chr7', 0, 56000000)\n",
    "\n",
    "50kb\n",
    " - ('chr6',125200000, 145200000)\n",
    "\n",
    "10kb assembly on- and off-diagonal\n",
    " - on diagonal #1: ('chr6', 129100000, 130500000)\n",
    " - on diagonal #2: ('chr6', 136100000, 138200000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93f6dd9b-ab25-4fcd-a25b-26d44a496006",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bws\n",
    "# mycmap = mpl.colors.LinearSegmentedColormap.from_list(\n",
    "#     name=\"myfall\",\n",
    "#     colors=[ mpl.colors.to_hex(c) for c in plotting.PALETTES[\"fall\"] ],\n",
    "#     N=2**16,\n",
    "# )\n",
    "# mycmap.set_bad('white',.5)\n",
    "# mycmap\n",
    "_chr6_len = bioframe.fetch_chromsizes(\"hg38\")[\"chr6\"]\n",
    "# region1_250 = ('chr6', 105000000, _chr6_len)\n",
    "# region2_250 = ('chr7', 0, 56000000)\n",
    "# region1_50 = ('chr6', 125200000, 145200000)\n",
    "region1_10 = ('chr6', 129100000, 130500000)\n",
    "region2_10 = ('chr6', 136100000, 138200000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78f29d9f-2e2f-47ce-9685-7982efbbb75f",
   "metadata": {},
   "outputs": [],
   "source": [
    "_chr6_len = bioframe.fetch_chromsizes(\"hg38\")[\"chr6\"]\n",
    "# region1_250 = ('chr6', 105000000+25_000_000, _chr6_len)\n",
    "# region2_250 = ('chr7', 0, 56000000-35_000_000)\n",
    "# # ...\n",
    "# region1_50 = ('chr6', 125200000+1_500_000, 145200000-4_500_000)\n",
    "# # ...\n",
    "region1_10 = ('chr6', 129100000-30_000, 130500000+30_000)\n",
    "region2_10 = ('chr6', 136100000-45_000, 138200000+45_000)\n",
    "region1_10v = ('chr6', 129100000-30_000, 130500000-730_000+30_000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3a74556-0596-4481-8583-557c25a0a10d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cooltools.lib import numutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73bcff6d-5ea1-46d2-8f4b-e99c04046e62",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# def get_hm(\n",
    "#     clr,\n",
    "#     _idx_from,\n",
    "#     _idx_to,\n",
    "#     _jdx_from=None,\n",
    "#     _jdx_to=None,\n",
    "#     vmin=0.0002,\n",
    "#     vmax=0.0125,\n",
    "# ):\n",
    "#     #\n",
    "#     _jdx_from = _idx_from if _jdx_from is None else _jdx_from\n",
    "#     _jdx_to = _idx_to if _jdx_to is None else _jdx_to\n",
    "#     _mat = xr.DataArray(\n",
    "#         clr.matrix()[_idx_from:_idx_to, _jdx_from:_jdx_to]\n",
    "#     )\n",
    "#     return tf.shade(\n",
    "#         _mat,\n",
    "#         cmap = mycmap,\n",
    "#         # how = lambda x,y: np.power(x, 1/3.),\n",
    "#         how = lambda x,y: np.log10(x),\n",
    "#         span = (vmin, vmax),\n",
    "#     ).to_pil(origin=\"upper\")\n",
    "\n",
    "\n",
    "# switch back to s\n",
    "def get_hm(\n",
    "    clr,\n",
    "    _idx_from,\n",
    "    _idx_to,\n",
    "    _jdx_from=None,\n",
    "    _jdx_to=None,\n",
    "    vmin=0.0002,\n",
    "    vmax=0.0125,\n",
    "):\n",
    "    #\n",
    "    _jdx_from = _idx_from if _jdx_from is None else _jdx_from\n",
    "    _jdx_to = _idx_to if _jdx_to is None else _jdx_to\n",
    "    _mat = clr.matrix()[_idx_from:_idx_to, _jdx_from:_jdx_to]\n",
    "    # need to NaN-out 2 diagonals ...\n",
    "    if _idx_from == _jdx_from:\n",
    "        for i in [-1,0,1]:\n",
    "            _mat = numutils.fill_diag(_mat, np.nan, i)\n",
    "        # ...\n",
    "    _imshow_kwargs = dict(\n",
    "        norm = LogNorm(vmin=vmin, vmax=vmax),\n",
    "        cmap = mycmap,\n",
    "        interpolation=\"nearest\",\n",
    "    )\n",
    "    return _mat, _imshow_kwargs\n",
    "\n",
    "\n",
    "# vlims_dict = {\n",
    "#     \"250\": dict(vmin=0.000005, vmax=0.002),\n",
    "#     \"50\": dict(vmin=0.0001, vmax=0.01),\n",
    "#     \"10_1\": dict(vmin=0.0002, vmax=0.0125),\n",
    "#     \"10_2\": dict(vmin=0.0001, vmax=0.0025),\n",
    "# }\n",
    "\n",
    "vlims_dict = {\n",
    "    \"250\": dict(vmin=0.00001, vmax=0.003),\n",
    "    \"50\": dict(vmin=0.00004, vmax=0.025),\n",
    "    # \"10_1\": dict(vmin=0.00006, vmax=0.06),\n",
    "    # \"10_2\": dict(vmin=0.00006, vmax=0.06),\n",
    "    \"10_1\": dict(vmin=0.0001, vmax=0.01),\n",
    "    \"10_2\": dict(vmin=0.0001, vmax=0.01),\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43792609-0eab-4a6b-8979-6d5055648c1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import matplotlib.pyplot as plt\n",
    "import bbi\n",
    "import matplotlib.lines as lines\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.patches import ConnectionPatch, Rectangle\n",
    "from mpl_toolkits.axes_grid1 import Divider, Size\n",
    "from mpl_toolkits.axes_grid1.inset_locator import BboxConnector\n",
    "\n",
    "# manual colorbar example ...\n",
    "# #\n",
    "# cax = divider.append_axes(\"right\", size=0.1, pad=0.05)\n",
    "# plt.colorbar(\n",
    "#     cm.ScalarMappable(norm=PowerNorm(1/3., vmin=vmin, vmax=vmax), cmap=mycmap),\n",
    "#     cax=cax,\n",
    "# )\n",
    "\n",
    "# telo_clrs250.keys()\n",
    "# ! ls -l ev_bigwig\n",
    "# fig.add_axes((left, bottom, width, height))\n",
    "\n",
    "\n",
    "# in a specific region, and exposing importnat plotting parameters\n",
    "def rectangles_around_dots_ww(dots_bins_df, the_tile, loc=\"upper\", lw=1, ec=\"cyan\", fc=\"none\", halo=30_000, ext_width=0):\n",
    "    rectangle_kwargs = dict(lw=lw, ec=ec, fc=fc)\n",
    "    # parse the tile\n",
    "    _, tspan1, tspan2 = the_tile\n",
    "    # select only visible \"boxes\" :\n",
    "    _the_dots = dots_bins_df \\\n",
    "        .query(\"\"\"(@tspan1[0] - @halo < bin1_id < @tspan1[1] + @halo) & \\\n",
    "                  (@tspan2[0] - @halo < bin2_id < @tspan2[1] + @halo) \"\"\") \\\n",
    "        .eval(\"\"\"\n",
    "                b1 = bin1_id - @tspan1[0] - @ext_width - 0.5\n",
    "                b2 = bin2_id - @tspan2[0] - @ext_width - 0.5\n",
    "                bin1_width = bin1_width + 2*@ext_width\n",
    "                bin2_width = bin2_width + 2*@ext_width\n",
    "            \"\"\")\n",
    "    print(f\"{len(_the_dots)} pixels are visible out of {len(dots_bins_df)} ...\")\n",
    "    for b1, b2, w1, w2 in _the_dots[[\"b1\", \"b2\", \"bin1_width\", \"bin2_width\"]].itertuples(index=False):\n",
    "        if loc == \"upper\":\n",
    "            yield patches.Rectangle((b2, b1), w2+1, w1+1, **rectangle_kwargs)\n",
    "        elif loc == \"lower\":\n",
    "            yield patches.Rectangle((b1, b2), w1+1, w2+1, **rectangle_kwargs)\n",
    "        else:\n",
    "            raise ValueError(\"loc has to be uppper or lower\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def rectangles_at_anchors_vert(anchors_df, region_extent, ax, halo=30_000, ext_width=0, **span_kwargs):\n",
    "    # rectangle_kwargs = dict(lw=lw, ec=ec, fc=fc)\n",
    "    # parse the tile\n",
    "    _bin_start, _bin_end = region_extent\n",
    "    # select only visible \"boxes\" :\n",
    "    _the_anchors = anchors_df \\\n",
    "        .query(\"\"\"(@_bin_start - @halo < bin_id < @_bin_end + @halo)\"\"\") \\\n",
    "        .eval(\"\"\"\n",
    "                b = bin_id - @_bin_start - @ext_width - 0.5\n",
    "                b_width = bin_width + 2*@ext_width\n",
    "            \"\"\")\n",
    "    print(f\"{len(_the_anchors)} anchors are visible out of {len(anchors_df)} ...\")\n",
    "    for b, w in _the_anchors[[\"b\", \"b_width\"]].itertuples(index=False):\n",
    "        ax.axhspan(b, b+w, **span_kwargs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "feaae19c-b3ee-416f-b8bd-056349c7abb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "cre_to_code = {\n",
    "    'pls':6,\n",
    "    'pels':5,\n",
    "    'nels':5,\n",
    "    'dels':4,\n",
    "    'openk4':3,\n",
    "    'ctcf':2,\n",
    "    'justopen':1,\n",
    "}\n",
    "\n",
    "_cats = pd.CategoricalDtype(categories=list(range(6+1)), ordered=True)\n",
    "\n",
    "cre_fnames = {\n",
    "    \"atac95_rest95\" : \"./bigbed/all_cres_encodestyle_pMunion_95atac_95rest.bed\",  # control sorta, like before\n",
    "    # \"atac90_rest90\" : \"./bigbed/all_cres_encodestyle_pMunion_90atac_90rest.bed\",\n",
    "    # \"atac70_rest70\" : \"./bigbed/all_cres_encodestyle_pMunion_70atac_70rest.bed\",\n",
    "    # \"atac70_rest95\" : \"./bigbed/all_cres_encodestyle_pMunion_70atac_95rest.bed\",\n",
    "    # \"atac70_rest90\" : \"./bigbed/all_cres_encodestyle_pMunion_70atac_90rest.bed\",\n",
    "    # \"atac75_rest95\" : \"./bigbed/all_cres_encodestyle_pMunion_75atac_95rest.bed\",\n",
    "    # \"atac75_rest90\" : \"./bigbed/all_cres_encodestyle_pMunion_75atac_90rest.bed\",\n",
    "}\n",
    "\n",
    "_status_query = \"(status_ == 0)|(status_ == 5)|(status_ == 4)|(status_ == 6)|(status_ == 7)\"\n",
    "\n",
    "binned_cre10_dfs = {}\n",
    "for name, cre_fname in cre_fnames.items():\n",
    "    df = pd.read_table(cre_fname, usecols=['chrom', 'start', 'end', 'name'])\n",
    "    print(f\"read {name=} with {len(df)} items ...\")\n",
    "    # # rename nels -> pels\n",
    "    # df = df.replace({'name': {\"nels\": \"pels\"}})\n",
    "    # sort by coordinate\n",
    "    df = bioframe.sort_bedframe(df, view_df=hg38_arms)\n",
    "    df[\"status\"] = df[\"name\"].map(cre_to_code)\n",
    "\n",
    "    # now annotate bins with the anchors @10kb  ...\n",
    "    _bin_cres = bioframe.overlap(clr_bins10, df)\n",
    "    _bin_cres = _bin_cres.drop(columns=[\"chrom_\",\"start_\",\"end_\"])\n",
    "    _bin_cres[\"status_\"] = _bin_cres[\"status_\"].fillna(0).astype(int)\n",
    "    # assign cCREs per bin according to the hierarchy ...\n",
    "    binned_cres = _bin_cres.groupby([\"chrom\",\"start\",\"end\"], observed=True)[\"status_\"].max().reset_index()\n",
    "    binned_cres[\"status_\"] = binned_cres[\"status_\"].astype(_cats)\n",
    "    # value_counts of annotated bins before gene assignment ...\n",
    "    print(\"@10kb\\n\")\n",
    "    display(binned_cres[\"status_\"].value_counts().sort_index())\n",
    "    print(\"\\n\")\n",
    "    #\n",
    "    binned_cre10_dfs[name] = binned_cres.astype({ \"chrom\" : str })\n",
    "\n",
    "\n",
    "ticklabels = [\n",
    "    \"PLS\",\n",
    "    \"pELS\",\n",
    "    \"dELS\",\n",
    "    \"opK4\",\n",
    "    \"CTCF\",\n",
    "    \"open\",\n",
    "    \"none\",\n",
    "]\n",
    "\n",
    "\n",
    "cre_cmap = {\n",
    "    6: \"darkred\",  #\"PLS\",\n",
    "    5: \"red\",  #\"pELS\",\n",
    "    4: \"orangered\",  #\"dELS\",\n",
    "    3: \"pink\",  #\"opK4\",\n",
    "    2: \"blue\",  #\"CTCF\",\n",
    "    1: \"#ffee99\",  #\"open\",\n",
    "    # \"none\",\n",
    "}\n",
    "# # categroies are going to be 1,2,3,4,5,6\n",
    "# # 0 - everything else ...\n",
    "_none_color = \"#D9E2EF\"\n",
    "\n",
    "\n",
    "_working_cres10 = binned_cre10_dfs['atac95_rest95']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eefdcfb3-ebb1-49d5-a4ff-258063e9e4d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "_dfcre_reg1 = bioframe.select(_working_cres10, region1_10).reset_index(drop=True)\n",
    "_dfcre_reg2 = bioframe.select(_working_cres10, region2_10).reset_index(drop=True)\n",
    "\n",
    "_dfcre_reg1 = _dfcre_reg1.query(\"status_ > 0\").reset_index(drop=True)\n",
    "_dfcre_reg2 = _dfcre_reg2.query(\"status_ > 0\").reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d9bfcdb-76c1-4e73-8f66-7cd5165db4f0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6ff41cc-1b62-4cc7-9112-15e6a08fbd49",
   "metadata": {},
   "outputs": [],
   "source": [
    "bws_vlim['mG.atac'][\"vmin\"] = 1\n",
    "bws_vlim['mG.atac'][\"vmax\"] = 100\n",
    "\n",
    "bws_vlim['H3K27ac'][\"vmin\"] = 10\n",
    "bws_vlim['H3K27ac'][\"vmax\"] = 1000\n",
    "\n",
    "bws_vlim['H3K4me3'][\"vmin\"] = 10\n",
    "bws_vlim['H3K4me3'][\"vmax\"] = 1000\n",
    "\n",
    "# inactive genes ...\n",
    "fwd_inact_args = fwd_args.copy()\n",
    "fwd_inact_args[\"color\"] = \"lightgrey\"\n",
    "\n",
    "rev_inact_args = rev_args.copy()\n",
    "rev_inact_args[\"color\"] = \"lightgrey\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d5ca258-d354-406b-ac19-74665f1bb195",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "margin = 0.2\n",
    "matw = 1.333\n",
    "cbarh = 0.07\n",
    "cbarw = 0.7*matw\n",
    "\n",
    "\n",
    "_region12_factor = (region2_10[2] - region2_10[1])/(region1_10[2] - region1_10[1])\n",
    "_region1v_factor = (region1_10v[2] - region1_10v[1])/(region1_10[2] - region1_10[1])\n",
    "# The first items are for padding and the second items are for the axes, sizes are in inch.\n",
    "h = [\n",
    "    Size.Fixed(margin), Size.Fixed(matw), Size.Fixed(0.2*margin), Size.Fixed(matw*_region12_factor),\n",
    "]\n",
    "# goes from bottom to the top ...\n",
    "v = [\n",
    "    Size.Fixed(margin), Size.Fixed(matw*0.075),\n",
    "    Size.Fixed(0.2*margin), Size.Fixed(matw*0.13), Size.Fixed(matw*0.13), Size.Fixed(matw*0.13), Size.Fixed(matw*0.13),\n",
    "    Size.Fixed(0.2*margin), Size.Fixed(matw*0.075),\n",
    "    Size.Fixed(0.2*margin), Size.Fixed(matw*_region1v_factor), Size.Fixed(margin)\n",
    "]\n",
    "\n",
    "# set figsize based on the tiling provided ...\n",
    "fig_width = sum(_h.fixed_size for _h in h)\n",
    "fig_height = sum(_v.fixed_size for _v in v)\n",
    "fig = plt.figure(\n",
    "    figsize=(fig_width, fig_height),\n",
    "    # facecolor='lightblue'\n",
    ")\n",
    "print(f\"figure size {fig_width=} {fig_height=}\")\n",
    "\n",
    "divider = Divider(fig, (0, 0, 1, 1), h, v, aspect=False)\n",
    "axs = {}\n",
    "# # 10 kb region1 and region2 ....\n",
    "axs[\"hmap_reg1_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1, ny=10))\n",
    "axs[\"ids_reg1_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1, ny=8))\n",
    "axs[\"atac_reg1_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1, ny=6))\n",
    "axs[\"k27ac_reg1_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1, ny=5))\n",
    "axs[\"k4me3_reg1_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1, ny=4))\n",
    "axs[\"genes_reg1_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1, ny=3))\n",
    "axs[\"cres_reg1_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1, ny=1))\n",
    "\n",
    "\n",
    "axs[\"hmap_reg2_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3, ny=10))\n",
    "axs[\"ids_reg2_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3, ny=8))\n",
    "axs[\"atac_reg2_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3, ny=6))\n",
    "axs[\"k27ac_reg2_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3, ny=5))\n",
    "axs[\"k4me3_reg2_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3, ny=4))\n",
    "axs[\"genes_reg2_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3, ny=3))\n",
    "axs[\"cres_reg2_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3, ny=1))\n",
    "\n",
    "# axs[\"hmap1_reg2_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3, ny=1+2))\n",
    "# axs[\"track_reg2_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3, ny=3+2))\n",
    "# axs[\"hmap2_reg2_bs10\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3, ny=5+2))\n",
    "\n",
    "\n",
    "########################################\n",
    "for k, ax in axs.items():\n",
    "    ax.xaxis.set_visible(False)\n",
    "    ax.yaxis.set_visible(False)\n",
    "\n",
    "_psample = \"p5hR1R2\"\n",
    "clr10 = telo_clrs10[_psample]\n",
    "\n",
    "# # heatmap and track for 250 kb\n",
    "# _i1_250, _i10_250 = clr250.extent(region1_250)\n",
    "# _i20_250, _i2_250 = clr250.extent(region2_250)\n",
    "# _track_size_250 = _i2_250 - _i1_250\n",
    "# mtrack1_250 = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_msample}.10kb.bw\",region1_250[0],region1_250[1],region1_250[2], bins=_i10_250-_i1_250 ))\n",
    "# mtrack2_250 = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_msample}.10kb.bw\",region2_250[0],region2_250[1],region2_250[2], bins=_i2_250-_i20_250 ))\n",
    "# ptrack1_250 = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_psample}.10kb.bw\",region1_250[0],region1_250[1],region1_250[2], bins=_i10_250-_i1_250 ))\n",
    "# ptrack2_250 = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_psample}.10kb.bw\",region2_250[0],region2_250[1],region2_250[2], bins=_i2_250-_i20_250 ))\n",
    "\n",
    "# # heatmap and track @ 50 kb ...\n",
    "# _i1_50, _i2_50 = clr50.extent(region1_50)\n",
    "# _track_size_50 = _i2_50 - _i1_50\n",
    "# mtrack1_50 = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_msample}.10kb.bw\",region1_50[0],region1_50[1],region1_50[2], bins=_i2_50-_i1_50 ))\n",
    "# ptrack1_50 = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_psample}.10kb.bw\",region1_50[0],region1_50[1],region1_50[2], bins=_i2_50-_i1_50 ))\n",
    "\n",
    "# and region1 and 2 @ 10kb ...\n",
    "_i1_10, _i2_10 = clr10.extent(region1_10)\n",
    "_j1_10, _j2_10 = clr10.extent(region2_10)\n",
    "_track_size1_10 = _i2_10 - _i1_10\n",
    "_track_size2_10 = _j2_10 - _j1_10\n",
    "# vertical side ...\n",
    "_i1_10v, _i2_10v = clr10.extent(region1_10v)\n",
    "# mtrack1_10 = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_msample}.10kb.bw\",region1_10[0],region1_10[1],region1_10[2], bins=_i2_10-_i1_10 ))\n",
    "# mtrack2_10 = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_msample}.10kb.bw\",region2_10[0],region2_10[1],region2_10[2], bins=_j2_10-_j1_10 ))\n",
    "# ptrack1_10 = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_psample}.10kb.bw\",region1_10[0],region1_10[1],region1_10[2], bins=_i2_10-_i1_10 ))\n",
    "# ptrack2_10 = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_psample}.10kb.bw\",region2_10[0],region2_10[1],region2_10[2], bins=_j2_10-_j1_10 ))\n",
    "\n",
    "# for ax_tag, ax in axs.items():\n",
    "#     if \"track\" in ax_tag:\n",
    "#         ax.axhline(0, color=\"grey\", lw=0.5)\n",
    "#         ax.set_ylim(-2, 2)\n",
    "\n",
    "# 10 ...\n",
    "hm, hmkwargs = get_hm(clr10, _i1_10v, _i2_10v, _i1_10, _i2_10, **vlims_dict[\"10_1\"])\n",
    "axs[\"hmap_reg1_bs10\"].imshow(hm, **hmkwargs)\n",
    "print(f\"1st 510kb heatmap is {_track_size1_10} worth of pixels\")\n",
    "\n",
    "# and region2 @ 10kb ...\n",
    "hm, hmkwargs = get_hm(clr10, _i1_10v, _i2_10v, _j1_10, _j2_10, **vlims_dict[\"10_2\"])\n",
    "axs[\"hmap_reg2_bs10\"].imshow(hm, **hmkwargs)\n",
    "print(f\"2nd 10kb heatmap is {_track_size2_10} worth of pixels\")\n",
    "\n",
    "\n",
    "\n",
    "_track_factor=10_000\n",
    "_bin_factor = 300/(region2_10[2] - region2_10[1])\n",
    "\n",
    "\n",
    "#do tracks in a cycle ...\n",
    "for _track, _bin_factor, _vlims, _ax1, _ax2, _color in zip(\n",
    "    [bws['mG.atac'], bws['H3K27ac'], bws['H3K4me3']],\n",
    "    [2*_bin_factor, 2*_bin_factor*0.75, 2*_bin_factor*0.666],\n",
    "    [(0.03, 2), (0.001, 75), (0.001, 35)],\n",
    "    [axs[\"atac_reg1_bs10\"], axs[\"k27ac_reg1_bs10\"], axs[\"k4me3_reg1_bs10\"]],\n",
    "    [axs[\"atac_reg2_bs10\"], axs[\"k27ac_reg2_bs10\"], axs[\"k4me3_reg2_bs10\"]],\n",
    "    [\"black\", \"red\", \"darkred\"],\n",
    "):\n",
    "    _bins1 = int( _bin_factor*( region1_10[2] - region1_10[1] ) )\n",
    "    _bins2 = int( _bin_factor*( region2_10[2] - region2_10[1] ) )\n",
    "    track1_10 = bbi.fetch(_track, region1_10[0], region1_10[1], region1_10[2], bins=_bins1, summary=\"mean\", missing=0.0, oob=0.0)\n",
    "    track2_10 = bbi.fetch(_track, region2_10[0], region2_10[1], region2_10[2], bins=_bins2, summary=\"mean\", missing=0.0, oob=0.0)\n",
    "    # _ax1.bar(\n",
    "    #     x=np.arange(len(track1_10)),\n",
    "    #     height=track1_10,\n",
    "    #     width=1.2,\n",
    "    #     log=False,\n",
    "    #     color=_color,\n",
    "    # )\n",
    "    # _ax1.set_xlim(0, _bins1)\n",
    "    # _ax1.set_ylim(*_vlims)\n",
    "    # _ax2.bar(\n",
    "    #     x=np.arange(len(track2_10)),\n",
    "    #     height=track2_10,\n",
    "    #     width=1.2,\n",
    "    #     log=False,\n",
    "    #     color=_color,\n",
    "    # )\n",
    "    # _ax2.set_xlim(0, _bins2)\n",
    "    # _ax2.set_ylim(*_vlims)\n",
    "    _ax1.fill_between(\n",
    "        x=np.arange(len(track1_10)),\n",
    "        y1=0,\n",
    "        y2=track1_10,\n",
    "        linewidth=0.5,\n",
    "        # antialiased=True,\n",
    "        # log=False,\n",
    "        color=_color,\n",
    "    )\n",
    "    _ax1.set_xlim(0, _bins1)\n",
    "    _ax1.set_ylim(*_vlims)\n",
    "    _ax2.fill_between(\n",
    "        x=np.arange(len(track2_10)),\n",
    "        y1=0,\n",
    "        y2=track2_10,\n",
    "        linewidth=0.5,\n",
    "        # antialiased=True,\n",
    "        # log=False,\n",
    "        color=_color,\n",
    "    )\n",
    "    _ax2.set_xlim(0, _bins2)\n",
    "    _ax2.set_ylim(*_vlims)\n",
    "\n",
    "##############################################\n",
    "# add ID-ID ...\n",
    "##############################################\n",
    "\n",
    "# ...\n",
    "_big_boxes_kwargs = dict(halo=100, lw=0.5, ec='black')\n",
    "# ...\n",
    "ax = axs[\"hmap_reg1_bs10\"]\n",
    "axid = axs[\"ids_reg1_bs10\"]\n",
    "axid.set_xlim(0-0.5,_track_size1_10-0.5)\n",
    "# axs[\"hmap_reg2_bs10\"]\n",
    "# draw boxes around clustered pixels ...\n",
    "for box in rectangles_around_dots_ww(\n",
    "    _idid,\n",
    "    (region1_10[0], (_i1_10v, _i2_10v), (_i1_10, _i2_10)),\n",
    "    ext_width=0,\n",
    "    **_big_boxes_kwargs,\n",
    "):\n",
    "    ax.add_patch(box)\n",
    "    x0, y0, dx, dy = box.get_bbox().bounds\n",
    "    axid.axvspan(x0, x0+dx, facecolor='black', alpha=1)\n",
    "\n",
    "\n",
    "ax = axs[\"hmap_reg2_bs10\"]\n",
    "axid = axs[\"ids_reg2_bs10\"]\n",
    "axid.set_xlim(0-0.5,_track_size2_10-0.5)\n",
    "# draw boxes around clustered pixels ...\n",
    "for box in rectangles_around_dots_ww(\n",
    "    _idid,\n",
    "    (region1_10[0], (_i1_10v, _i2_10v), (_j1_10, _j2_10)),\n",
    "    ext_width=0,\n",
    "    **_big_boxes_kwargs,\n",
    "):\n",
    "    ax.add_patch(box)\n",
    "    x0, y0, dx, dy = box.get_bbox().bounds\n",
    "    axid.axvspan(x0, x0+dx, facecolor='black', alpha=1)\n",
    "\n",
    "# ########################\n",
    "# Plots genes !\n",
    "# region 1\n",
    "_start, _end = region1_10[1], region1_10[2]\n",
    "# _start, _end = region1_10[1]-200_000, region1_10[2]+200_000\n",
    "_ax = axs[\"genes_reg1_bs10\"]\n",
    "# _gfs_fwd = GraphicRecord(\n",
    "#     sequence_length=(_end-_start+1),\n",
    "#     features=[\n",
    "#         GraphicFeature(_row.start, _row.end, **fwd_inact_args) for _row in \\\n",
    "#         bioframe.merge(\n",
    "#             bioframe.expand(bioframe.select(_fwd_inact, region1_10), pad=8_000),\n",
    "#             min_dist = 20_000,\n",
    "#         ).itertuples(index=False)\n",
    "#     ],\n",
    "# )\n",
    "_gfs_rev = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **rev_inact_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_rev_inact, region1_10), pad=8_000),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_fwd.plot(ax=_ax, x_lim=(_start, _end), level_offset=-0.45, **plotargs)\n",
    "_gfs_rev.plot(ax=_ax, x_lim=(_start, _end), level_offset=0.45, **plotargs)\n",
    "# ...\n",
    "_gfs_fwd = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **fwd_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_fwd_act, region1_10), pad=8_000),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_rev = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **rev_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_rev_act, region1_10), pad=8_000),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_fwd.plot(ax=_ax, x_lim=(_start, _end), level_offset=-0.45, **plotargs)\n",
    "_gfs_rev.plot(ax=_ax, x_lim=(_start, _end), level_offset=0.45, **plotargs)\n",
    "_ax.set_ylim(-1,1)\n",
    "_ax.set_axis_on()\n",
    "\n",
    "\n",
    "\n",
    "# ########################\n",
    "# Plots genes !\n",
    "# region 2\n",
    "_start, _end = region2_10[1], region2_10[2]\n",
    "_ax = axs[\"genes_reg2_bs10\"]\n",
    "_gfs_fwd = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **fwd_inact_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_fwd_inact, region2_10), pad=8_000),\n",
    "            min_dist=20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_rev = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **rev_inact_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_rev_inact, region2_10), pad=8_000),\n",
    "            min_dist=20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_fwd.plot(ax=_ax, x_lim=(_start, _end), level_offset=-0.45, **plotargs)\n",
    "_gfs_rev.plot(ax=_ax, x_lim=(_start, _end), level_offset=0.45, **plotargs)\n",
    "# ...\n",
    "_gfs_fwd = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **fwd_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_fwd_act, region2_10), pad=8_000),\n",
    "            min_dist=20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_rev = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **rev_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_rev_act, region2_10), pad=8_000),\n",
    "            min_dist=20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_fwd.plot(ax=_ax, x_lim=(_start, _end), level_offset=-0.45, **plotargs)\n",
    "_gfs_rev.plot(ax=_ax, x_lim=(_start, _end), level_offset=0.45, **plotargs)\n",
    "\n",
    "_ax.set_ylim(-1,1)\n",
    "_ax.set_axis_on()\n",
    "\n",
    "\n",
    "# axs[\"cres_reg1_bs10\"]\n",
    "# axs[\"cres_reg2_bs10\"]\n",
    "_reg1_start = region1_10[1]\n",
    "_reg2_start = region2_10[1]\n",
    "_reg1_len = region1_10[2] - _reg1_start\n",
    "_reg2_len = region2_10[2] - _reg2_start\n",
    "\n",
    "# draw CREs here ...\n",
    "for _chrom, _start, _end, _status_ in _dfcre_reg1.itertuples(index=False):\n",
    "    axs[\"cres_reg1_bs10\"].axvspan(\n",
    "        _start-_reg1_start,\n",
    "        _end-_reg1_start,\n",
    "        facecolor=cre_cmap[_status_],\n",
    "        lw=0.0,\n",
    "        ec=None,\n",
    "    )\n",
    "# draw CREs here ...\n",
    "for _chrom, _start, _end, _status_ in _dfcre_reg2.itertuples(index=False):\n",
    "    axs[\"cres_reg2_bs10\"].axvspan(\n",
    "        _start-_reg2_start,\n",
    "        _end-_reg2_start,\n",
    "        facecolor=cre_cmap[_status_],\n",
    "        lw=0.0,\n",
    "        ec=None,\n",
    "    )\n",
    "\n",
    "axs[\"cres_reg1_bs10\"].set_xlim(0,_reg1_len)\n",
    "axs[\"cres_reg2_bs10\"].set_xlim(0,_reg2_len)\n",
    "\n",
    "axs[\"cres_reg1_bs10\"].set_facecolor(_none_color)\n",
    "axs[\"cres_reg2_bs10\"].set_facecolor(_none_color)\n",
    "\n",
    "\n",
    "plt.savefig(\"Fig2B.pdf\", dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "518d2083-7d97-48ba-8e2d-9d3a9ff87917",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdfa659d-91be-49fd-9ab9-279468a296bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ecfc422-26f3-49a1-aa66-3609aa0b2a31",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93c948ef-7026-4144-b1a2-b78f2c108c5b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "baf1b477-7d76-42ec-9d94-9b9af213776f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "755ba66b-9935-4fcc-838e-f1ba4528508a",
   "metadata": {},
   "source": [
    "# Figure 5 heatmap with tracks and annotations ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5ee4691-ec4a-4a57-9d46-03c144a29e0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# switch back to s\n",
    "def get_hm_double(\n",
    "    clr_up,\n",
    "    clr_down,\n",
    "    _idx_from,\n",
    "    _idx_to,\n",
    "    _jdx_from=None,\n",
    "    _jdx_to=None,\n",
    "    vmin=0.0002,\n",
    "    vmax=0.0125,\n",
    "):\n",
    "    #\n",
    "    _jdx_from = _idx_from if _jdx_from is None else _jdx_from\n",
    "    _jdx_to = _idx_to if _jdx_to is None else _jdx_to\n",
    "    _mat_up = clr_up.matrix()[_idx_from:_idx_to, _jdx_from:_jdx_to]\n",
    "    _mat_down = clr_down.matrix()[_idx_from:_idx_to, _jdx_from:_jdx_to]\n",
    "    _mat = np.tril(_mat_down)+np.triu(_mat_up)\n",
    "    # need to NaN-out 2 diagonals ...\n",
    "    if _idx_from == _jdx_from:\n",
    "        for i in [-1,0,1]:\n",
    "            _mat = numutils.fill_diag(_mat, np.nan, i)\n",
    "        # ...\n",
    "    _imshow_kwargs = dict(\n",
    "        norm = LogNorm(vmin=vmin, vmax=vmax),\n",
    "        cmap = mycmap,\n",
    "        interpolation=\"nearest\",\n",
    "    )\n",
    "    return _mat, _imshow_kwargs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab5254ba-1761-4eb3-b59b-665e27972ddf",
   "metadata": {},
   "outputs": [],
   "source": [
    "_region5 = (\n",
    "    'chr7',\n",
    "    20_500_000,\n",
    "    28_500_000,\n",
    ")\n",
    "\n",
    "_region5_zoom = (\n",
    "    'chr7',\n",
    "    25_700_000,\n",
    "    28_000_000,\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "_msample = \"m5hR1R2\"\n",
    "clrM = telo_clrs25[_msample]\n",
    "\n",
    "_psample = \"p5hR1R2\"\n",
    "clrP = telo_clrs25[_psample]\n",
    "\n",
    "clrM_zoom = telo_clrs10[_msample]\n",
    "clrP_zoom = telo_clrs10[_psample]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4dbaa86-d056-458e-88f1-d91db63cf380",
   "metadata": {},
   "outputs": [],
   "source": [
    "# _region5\n",
    "\n",
    "\n",
    "\n",
    "# in a specific region, and exposing importnat plotting parameters\n",
    "def draw_ondiag_domains(bed_df, the_tile, lw=1, ec=\"cyan\", fc=\"none\", halo=30_000, ext_width=0):\n",
    "    rectangle_kwargs = dict(lw=lw, ec=ec, fc=fc)\n",
    "    # parse the tile\n",
    "    _, tspan1, tspan2 = the_tile\n",
    "    # select only visible \"boxes\" :\n",
    "    _the_dots = bed_df \\\n",
    "        .query(\"\"\"(@tspan1[0] - @halo < bin1_id < @tspan1[1] + @halo) & \\\n",
    "                  (@tspan2[0] - @halo < bin2_id < @tspan2[1] + @halo) \"\"\") \\\n",
    "        .eval(\"\"\"\n",
    "                b1 = bin1_id - @tspan1[0] - @ext_width\n",
    "                b2 = bin2_id - @tspan2[0] - @ext_width\n",
    "            \"\"\")\n",
    "    for b1, b2 in _the_dots[[\"b1\", \"b2\"]].itertuples(index=False):\n",
    "        yield patches.Rectangle((b1, b1), b2-b1+1, b2-b1+1, **rectangle_kwargs)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# in a specific region, and exposing importnat plotting parameters\n",
    "def rectangles_at_anchors(anchors_df, region_extent, ax, halo=30_000, ext_width=0, **span_kwargs):\n",
    "    # rectangle_kwargs = dict(lw=lw, ec=ec, fc=fc)\n",
    "    # parse the tile\n",
    "    _bin_start, _bin_end = region_extent\n",
    "    # select only visible \"boxes\" :\n",
    "    _the_anchors = anchors_df \\\n",
    "        .query(\"\"\"(@_bin_start - @halo < bin_id < @_bin_end + @halo)\"\"\") \\\n",
    "        .eval(\"\"\"\n",
    "                b = bin_id - @_bin_start - @ext_width - 0.5\n",
    "                b_width = bin_width + 2*@ext_width\n",
    "            \"\"\")\n",
    "    print(f\"{len(_the_anchors)} anchors are visible out of {len(anchors_df)} ...\")\n",
    "    for b, w in _the_anchors[[\"b\", \"b_width\"]].itertuples(index=False):\n",
    "        ax.axvspan(b, b+w, **span_kwargs)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# domains within selected region - turn it back to bedpe ...\n",
    "_domains_region = \\\n",
    "_domains.eval(\"\"\"\n",
    "    chrom1 = chrom\n",
    "    chrom2 = chrom\n",
    "    start1 = start\n",
    "    end1 = start + 10_000\n",
    "    start2 = end - 10_000\n",
    "    end2 = end\n",
    "\"\"\")[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2']]#, 'n_intervals']]\n",
    "\n",
    "# select domains in the region ...\n",
    "_domains_region = bioframe.select(\n",
    "    bioframe.select(_domains_region, _region5, cols=(\"chrom1\",\"start1\",\"end1\")),\n",
    "    _region5, cols=(\"chrom2\",\"start2\",\"end2\"),\n",
    ").reset_index(drop=True)\n",
    "_domains_region[\"bin1_id\"] = _domains_region[[\"chrom1\",\"start1\",\"end1\"]].apply(clrM_zoom.offset,axis=1,result_type=\"expand\")\n",
    "_domains_region[\"bin1_width\"] = _domains_region[[\"chrom1\",\"start1\",\"end1\"]].apply(clrM_zoom.extent,axis=1,result_type=\"expand\").apply(np.diff,axis=1,result_type=\"expand\")[0]\n",
    "_domains_region[\"bin2_id\"] = _domains_region[[\"chrom2\",\"start2\",\"end2\"]].apply(clrM_zoom.offset,axis=1,result_type=\"expand\")\n",
    "_domains_region[\"bin2_width\"] = _domains_region[[\"chrom2\",\"start2\",\"end2\"]].apply(clrM_zoom.extent,axis=1,result_type=\"expand\").apply(np.diff,axis=1,result_type=\"expand\")[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05c0f995-494d-49f9-be0a-46ced5ac0fa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# and region1 and 2 @ 10kb ...\n",
    "_i1_5, _i2_5 = clrM.extent(_region5)\n",
    "_i1_5z, _i2_5z = clrM_zoom.extent(_region5_zoom)\n",
    "\n",
    "_i1_5z_inset, _i2_5z_inset = clrM.extent(_region5_zoom)\n",
    "_inset_size = _i2_5z_inset - _i1_5z_inset\n",
    "_i1_5z_inset -= _i1_5\n",
    "\n",
    "patch_kwargs = dict(color=\"black\", fc=\"none\", lw=0.5)\n",
    "\n",
    "margin = 0.2\n",
    "matw = 1.333\n",
    "cbarh = 0.07\n",
    "cbarw = 0.7*matw\n",
    "mat_to_mcd = 0.125*margin\n",
    "chrom_to_mat = 0.3*margin\n",
    "\n",
    "# The first items are for padding and the second items are for the axes, sizes are in inch.\n",
    "h = [\n",
    "    Size.Fixed(margin), Size.Fixed(matw*0.1), Size.Fixed(chrom_to_mat), Size.Fixed(matw),  # 25kb hic\n",
    "    Size.Fixed(mat_to_mcd), Size.Fixed(cbarh),  # vertical MCDs ...\n",
    "    Size.Fixed(margin), Size.Fixed(matw),  # 10kb zoom in\n",
    "    Size.Fixed(mat_to_mcd), Size.Fixed(cbarh),  # MCDs and genes ...\n",
    "    Size.Fixed(margin)\n",
    "]\n",
    "# goes from bottom to the top ...\n",
    "v = [\n",
    "    Size.Fixed(margin),\n",
    "    Size.Fixed(cbarh), Size.Fixed(0.2*margin),\n",
    "    Size.Fixed(matw*0.11), Size.Fixed(matw*0.075),  # genes, 27ac\n",
    "    Size.Fixed(matw*0.1), Size.Fixed(0.2*margin), Size.Fixed(cbarh),  # EVs, MCDs\n",
    "    Size.Fixed(mat_to_mcd), Size.Fixed(matw),\n",
    "    Size.Fixed(margin)\n",
    "]\n",
    "\n",
    "# set figsize based on the tiling provided ...\n",
    "fig_width = sum(_h.fixed_size for _h in h)\n",
    "fig_height = sum(_v.fixed_size for _v in v)\n",
    "fig = plt.figure(\n",
    "    figsize=(fig_width, fig_height),\n",
    "    # facecolor='lightblue'\n",
    ")\n",
    "print(f\"figure size {fig_width=} {fig_height=}\")\n",
    "\n",
    "divider = Divider(fig, (0, 0, 1, 1), h, v, aspect=False)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "axs = {}\n",
    "# # 10 kb region1 and region2 ....\n",
    "axs[\"chrom\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1, ny=7+2))\n",
    "\n",
    "axs[\"hmap\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1+2, ny=7+2))\n",
    "axs[\"hmap_zoom\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=7+2))\n",
    "\n",
    "axs[\"genes\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1+2, ny=1+2))\n",
    "axs[\"k27ac\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1+2, ny=2+2))\n",
    "axs[\"evs\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1+2, ny=3+2))\n",
    "axs[\"mcd\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1+2, ny=5+2))\n",
    "axs[\"mcd_vert\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+2, ny=7+2))\n",
    "\n",
    "axs[\"genes_zoom\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=1+2))\n",
    "axs[\"k27ac_zoom\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=2+2))\n",
    "axs[\"evs_zoom\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=3+2))\n",
    "axs[\"mcd_zoom\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=5+2))\n",
    "axs[\"mcd_zoom_vert\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=5+4, ny=7+2))\n",
    "\n",
    "mtrack = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_msample}.{binsize25//1_000}kb.bw\",_region5[0],_region5[1],_region5[2], bins=_i2_5-_i1_5 ))\n",
    "mtrack_zoom = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_msample}.{binsize10//1_000}kb.bw\",_region5_zoom[0],_region5_zoom[1],_region5_zoom[2], bins=_i2_5z-_i1_5z ))\n",
    "ptrack = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_psample}.{binsize25//1_000}kb.bw\",_region5[0],_region5[1],_region5[2], bins=_i2_5-_i1_5 ))\n",
    "ptrack_zoom = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_psample}.{binsize10//1_000}kb.bw\",_region5_zoom[0],_region5_zoom[1],_region5_zoom[2], bins=_i2_5z-_i1_5z ))\n",
    "\n",
    "\n",
    "cbar = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1+2, ny=1))\n",
    "cbar_zoom = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=1))\n",
    "\n",
    "\n",
    "_mev_color = \"tab:blue\"\n",
    "_pev_color = \"tab:red\"\n",
    "\n",
    "\n",
    "########################################\n",
    "for k, ax in axs.items():\n",
    "    # ax.xaxis.set_visible(False)\n",
    "    # ax.yaxis.set_visible(True)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "\n",
    "\n",
    "# plot EV tracks ...\n",
    "axs[\"evs\"].axhline(0, lw=0.5, color=\"grey\")\n",
    "axs[\"evs\"].plot(mtrack, lw=0.7, color=_mev_color)\n",
    "axs[\"evs\"].plot(ptrack, lw=0.7, color=_pev_color)\n",
    "axs[\"evs\"].set_xlim(0, len(mtrack))\n",
    "\n",
    "\n",
    "axs[\"evs_zoom\"].axhline(0, lw=0.5, color=\"grey\")\n",
    "axs[\"evs_zoom\"].plot(mtrack_zoom, lw=0.7, color=_mev_color)\n",
    "axs[\"evs_zoom\"].plot(ptrack_zoom, lw=0.7, color=_pev_color)\n",
    "axs[\"evs_zoom\"].set_xlim(0, len(mtrack_zoom))\n",
    "\n",
    "\n",
    "hm, hmkwargs = get_hm_double(clrM, clrP, _i1_5, _i2_5, **vlims_dict[\"50\"])\n",
    "hmkwargs[\"interpolation\"] = \"antialiased\"\n",
    "hmkwargs[\"interpolation_stage\"] = \"rgba\"\n",
    "axs[\"hmap\"].imshow(hm, **hmkwargs)\n",
    "\n",
    "_rec = Rectangle(\n",
    "    (_i1_5z_inset, _i1_5z_inset),\n",
    "    _inset_size, _inset_size,\n",
    "    **patch_kwargs,\n",
    ")\n",
    "# draw those box and lines ...\n",
    "axs[\"hmap\"].add_patch(_rec)\n",
    "\n",
    "\n",
    "#################################\n",
    "# chrom schematics on both axes ...\n",
    "ax = axs[\"chrom\"]\n",
    "_7pl = arms_dict[\"chr7_p\"]\n",
    "_7ql = arms_dict[\"chr7_q\"]\n",
    "_regions_highlight = [(\"p\", _7pl-20, _7pl-29, \"dimgray\"), ]\n",
    "_ww = 0.1*(_7pl+_7ql)\n",
    "_7, _7hs = get_chrom_path_wpathes(\n",
    "    (0, 0),  # position of the centromere ...\n",
    "    parm=_7pl,  # p arm length\n",
    "    qarm=_7ql,  # q arm length\n",
    "    width=_ww,  # width\n",
    "    color=\"lightsteelblue\",\n",
    "    patches=_regions_highlight,\n",
    ")\n",
    "ax.set_axis_off()\n",
    "ax.add_artist(_7)\n",
    "for _h in _7hs:\n",
    "    ax.add_artist(_h)\n",
    "ax.set_xlim(-1.0*_ww, 1.0*_ww)\n",
    "ax.set_ylim(-(_7ql)-(0.11+0.1)*_ww, (_7pl)+(0.11+0.1)*_ww)\n",
    "##################################################\n",
    "ax_chrom = axs[\"chrom\"]\n",
    "ax = axs[\"hmap\"]\n",
    "con = ConnectionPatch(\n",
    "    xyA=(0.5*_ww, _7pl-20),\n",
    "    coordsA=ax_chrom.transData,\n",
    "    xyB=(0.0, 1.0),\n",
    "    coordsB=ax.transAxes,\n",
    "    **patch_kwargs,\n",
    ")\n",
    "ax.add_artist(con)\n",
    "con = ConnectionPatch(\n",
    "    xyA=(0.5*_ww, _7pl-29),\n",
    "    coordsA=ax_chrom.transData,\n",
    "    xyB=(0.0, 0.0),\n",
    "    coordsB=ax.transAxes,\n",
    "    **patch_kwargs,\n",
    ")\n",
    "ax.add_artist(con)\n",
    "# ##################################################\n",
    "\n",
    "# plot MCDs ...\n",
    "mcd_kwargs = dict(color=\"black\", alpha=1,ec=None)\n",
    "\n",
    "# doing that using clr.offset and clr.extent functionality:\n",
    "_anchors[\"bin_id\"] = _anchors[['chrom', 'peak_start', 'peak_end']].apply(clrM.offset, axis=1, result_type=\"expand\")\n",
    "_anchors[\"bin_width\"] = _anchors[['chrom', 'peak_start', 'peak_end']] \\\n",
    "    .apply(clrM.extent, axis=1, result_type=\"expand\") \\\n",
    "    .apply(np.diff, axis=1, result_type=\"expand\")[0]\n",
    "rectangles_at_anchors(_anchors, clrM.extent(_region5), axs[\"mcd\"], ext_width=-0.5, **mcd_kwargs)\n",
    "axs[\"mcd\"].set_xlim(0, _i2_5-_i1_5)\n",
    "\n",
    "# vertical ones ...\n",
    "rectangles_at_anchors_vert(_anchors, clrM.extent(_region5), axs[\"mcd_vert\"], ext_width=-0.5, **mcd_kwargs)\n",
    "axs[\"mcd_vert\"].set_ylim(0, _i2_5-_i1_5)\n",
    "axs[\"mcd_vert\"].invert_yaxis()\n",
    "\n",
    "# doing that using clr.offset and clr.extent functionality:\n",
    "_anchors[\"bin_id\"] = _anchors[['chrom', 'peak_start', 'peak_end']].apply(clrM_zoom.offset, axis=1, result_type=\"expand\")\n",
    "_anchors[\"bin_width\"] = _anchors[['chrom', 'peak_start', 'peak_end']] \\\n",
    "    .apply(clrM_zoom.extent, axis=1, result_type=\"expand\") \\\n",
    "    .apply(np.diff, axis=1, result_type=\"expand\")[0]\n",
    "rectangles_at_anchors(_anchors, clrM_zoom.extent(_region5_zoom), axs[\"mcd_zoom\"], ext_width=-0.5, **mcd_kwargs)\n",
    "axs[\"mcd_zoom\"].set_xlim(0, _i2_5z-_i1_5z)\n",
    "\n",
    "# vertical ones ...\n",
    "rectangles_at_anchors_vert(_anchors, clrM_zoom.extent(_region5_zoom), axs[\"mcd_zoom_vert\"], ext_width=-0.5, **mcd_kwargs)\n",
    "axs[\"mcd_zoom_vert\"].set_ylim(0, _i2_5z-_i1_5z)\n",
    "axs[\"mcd_zoom_vert\"].invert_yaxis()\n",
    "\n",
    "\n",
    "# ########################\n",
    "# Plots genes !\n",
    "# region 1\n",
    "_start, _end = _region5[1], _region5[2]\n",
    "_ax = axs[\"genes\"]\n",
    "_gfs_fwd = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **fwd_inact_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_fwd, _region5), pad=10_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_rev = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **rev_inact_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_rev, _region5), pad=10_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_fwd.plot(ax=_ax, x_lim=(_start, _end), level_offset=-0.45, **plotargs)\n",
    "_gfs_rev.plot(ax=_ax, x_lim=(_start, _end), level_offset=0.45, **plotargs)\n",
    "# ...\n",
    "_gfs_fwd = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **fwd_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_fwd_act, _region5), pad=10_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_rev = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **rev_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_rev_act, _region5), pad=10_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_fwd.plot(ax=_ax, x_lim=(_start, _end), level_offset=-0.45, **plotargs)\n",
    "_gfs_rev.plot(ax=_ax, x_lim=(_start, _end), level_offset=0.45, **plotargs)\n",
    "# ...\n",
    "_ax.set_ylim(-1,1)\n",
    "_ax.set_axis_on()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "_track_factor=10_000\n",
    "_bin_factor = 500/(_region5_zoom[2] - _region5_zoom[1])\n",
    "#do tracks in a cycle ...\n",
    "_track, _bin_factor, _vlims, _ax1, _ax2, _color = (\n",
    "    bws['H3K27ac'],\n",
    "    2*_bin_factor*0.75,\n",
    "    (0.001, 75),\n",
    "    axs[\"k27ac\"],\n",
    "    axs[\"k27ac_zoom\"],\n",
    "    \"red\",\n",
    ")\n",
    "\n",
    "_bins1 = int( _bin_factor*( _region5[2] - _region5[1] ) )\n",
    "_bins2 = int( _bin_factor*( _region5_zoom[2] - _region5_zoom[1] ) )\n",
    "track = bbi.fetch(_track, _region5[0], _region5[1], _region5[2], bins=_bins1, summary=\"mean\", missing=0.0, oob=0.0)\n",
    "track_zoom = bbi.fetch(_track, _region5_zoom[0], _region5_zoom[1], _region5_zoom[2], bins=_bins2, summary=\"mean\", missing=0.0, oob=0.0)\n",
    "_ax1.fill_between(\n",
    "    x=np.arange(len(track)),\n",
    "    y1=0,\n",
    "    y2=track,\n",
    "    linewidth=0.5,\n",
    "    # antialiased=True,\n",
    "    # log=False,\n",
    "    color=_color,\n",
    ")\n",
    "_ax1.set_xlim(0, _bins1)\n",
    "_ax1.set_ylim(*_vlims)\n",
    "_ax2.fill_between(\n",
    "    x=np.arange(len(track_zoom)),\n",
    "    y1=0,\n",
    "    y2=track_zoom,\n",
    "    linewidth=0.5,\n",
    "    # antialiased=True,\n",
    "    # log=False,\n",
    "    color=_color,\n",
    ")\n",
    "_ax2.set_xlim(0, _bins2)\n",
    "_ax2.set_ylim(*_vlims)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "hm, hmkwargs = get_hm_double(clrM_zoom, clrP_zoom, _i1_5z, _i2_5z, **vlims_dict[\"10_1\"])\n",
    "hmkwargs[\"interpolation\"] = \"antialiased\"\n",
    "hmkwargs[\"interpolation_stage\"] = \"rgba\"\n",
    "axs[\"hmap_zoom\"].imshow(hm, **hmkwargs)\n",
    "\n",
    "# draw domains on the zoomed in heatmap ...\n",
    "_big_boxes_kwargs = dict(lw=1., ec=\"green\", fc=\"none\", halo=100, ext_width=2)\n",
    "for box in draw_ondiag_domains(_domains_region, (\"...\", (_i1_5z, _i2_5z), (_i1_5z, _i2_5z)), **_big_boxes_kwargs):\n",
    "    axs[\"hmap_zoom\"].add_patch(box)\n",
    "\n",
    "\n",
    "# ########################\n",
    "# Plots genes !\n",
    "# region 1\n",
    "_start, _end = _region5_zoom[1], _region5_zoom[2]\n",
    "_ax = axs[\"genes_zoom\"]\n",
    "_gfs_fwd = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **fwd_inact_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_fwd, _region5_zoom), pad=8_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_rev = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **rev_inact_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_rev, _region5_zoom), pad=8_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_fwd.plot(ax=_ax, x_lim=(_start, _end), level_offset=-0.45, **plotargs)\n",
    "_gfs_rev.plot(ax=_ax, x_lim=(_start, _end), level_offset=0.45, **plotargs)\n",
    "# ...\n",
    "_gfs_fwd = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **fwd_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_fwd_act, _region5_zoom), pad=8_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_rev = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **rev_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_rev_act, _region5_zoom), pad=8_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_fwd.plot(ax=_ax, x_lim=(_start, _end), level_offset=-0.45, **plotargs)\n",
    "_gfs_rev.plot(ax=_ax, x_lim=(_start, _end), level_offset=0.45, **plotargs)\n",
    "_ax.set_ylim(-1,1)\n",
    "_ax.set_axis_on()\n",
    "\n",
    "\n",
    "\n",
    "axs[\"genes\"].axis(\"off\")\n",
    "axs[\"k27ac\"].axis(\"off\")\n",
    "axs[\"evs\"].axis(\"off\")\n",
    "\n",
    "axs[\"genes_zoom\"].axis(\"off\")\n",
    "axs[\"k27ac_zoom\"].axis(\"off\")\n",
    "axs[\"evs_zoom\"].axis(\"off\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# it's 25 actually - but whatever\n",
    "_vmin, _vmax = vlims_dict[\"50\"][\"vmin\"], vlims_dict[\"50\"][\"vmax\"]\n",
    "plt.colorbar(\n",
    "    # cm.ScalarMappable(norm=PowerNorm(1/3., vmin=_vmin, vmax=_vmax), cmap=mycmap),\n",
    "    cm.ScalarMappable(norm=LogNorm(vmin=_vmin, vmax=_vmax), cmap=mycmap),\n",
    "    cax=cbar,\n",
    "    orientation=\"horizontal\",\n",
    ")\n",
    "cbar.xaxis.set_visible(True)\n",
    "_ticks_power = range(\n",
    "    np.ceil(np.log10(_vmin)).astype(int),\n",
    "    np.floor(np.log10(_vmax)).astype(int)+1,\n",
    ")\n",
    "cbar.set_xticks([np.power(10.,_p) for _p in _ticks_power])\n",
    "cbar.set_xticklabels(['$\\mathdefault{10^{%i}}$' % l for l in _ticks_power])\n",
    "cbar.xaxis.set_tick_params(labelsize=6, pad=0.1, length=2)\n",
    "cbar.minorticks_off()\n",
    "\n",
    "\n",
    "# it's 25 actually - but whatever\n",
    "_vmin, _vmax = vlims_dict[\"10_1\"][\"vmin\"], vlims_dict[\"10_1\"][\"vmax\"]\n",
    "plt.colorbar(\n",
    "    # cm.ScalarMappable(norm=PowerNorm(1/3., vmin=_vmin, vmax=_vmax), cmap=mycmap),\n",
    "    cm.ScalarMappable(norm=LogNorm(vmin=_vmin, vmax=_vmax), cmap=mycmap),\n",
    "    cax=cbar_zoom,\n",
    "    orientation=\"horizontal\",\n",
    ")\n",
    "cbar_zoom.xaxis.set_visible(True)\n",
    "_ticks_power = range(\n",
    "    np.ceil(np.log10(_vmin)).astype(int),\n",
    "    np.floor(np.log10(_vmax)).astype(int)+1,\n",
    ")\n",
    "cbar_zoom.set_xticks([np.power(10.,_p) for _p in _ticks_power])\n",
    "cbar_zoom.set_xticklabels(['$\\mathdefault{10^{%i}}$' % l for l in _ticks_power])\n",
    "cbar_zoom.xaxis.set_tick_params(labelsize=6, pad=0.1, length=2)\n",
    "cbar_zoom.minorticks_off()\n",
    "\n",
    "\n",
    "\n",
    "plt.savefig(\"Fig5J.svg\", dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73895d2d-0699-4031-a692-4b63dfffd3f5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74893fb9-d70b-46a9-8dc8-f7d4a145cc36",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60eedee5-48bb-4740-9e5c-382b7989c25b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # domains within selected region - turn it back to bedpe ...\n",
    "# _domains_region = \\\n",
    "# _domains.eval(\"\"\"\n",
    "#     chrom1 = chrom\n",
    "#     chrom2 = chrom\n",
    "#     start1 = start\n",
    "#     end1 = start + 10_000\n",
    "#     start2 = end - 10_000\n",
    "#     end2 = end\n",
    "# \"\"\")[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2']]#, 'n_intervals']]\n",
    "\n",
    "# # select domains in the region ...\n",
    "# _domains_region = bioframe.select(\n",
    "#     bioframe.select(_domains_region, _region5, cols=(\"chrom1\",\"start1\",\"end1\")),\n",
    "#     _region5, cols=(\"chrom2\",\"start2\",\"end2\"),\n",
    "# ).reset_index(drop=True)\n",
    "# _domains_region[\"bin1_id\"] = _domains_region[[\"chrom1\",\"start1\",\"end1\"]].apply(clrM.offset,axis=1,result_type=\"expand\")\n",
    "# _domains_region[\"bin1_width\"] = _domains_region[[\"chrom1\",\"start1\",\"end1\"]].apply(clrM.extent,axis=1,result_type=\"expand\").apply(np.diff,axis=1,result_type=\"expand\")[0]\n",
    "# _domains_region[\"bin2_id\"] = _domains_region[[\"chrom2\",\"start2\",\"end2\"]].apply(clrM.offset,axis=1,result_type=\"expand\")\n",
    "# _domains_region[\"bin2_width\"] = _domains_region[[\"chrom2\",\"start2\",\"end2\"]].apply(clrM.extent,axis=1,result_type=\"expand\").apply(np.diff,axis=1,result_type=\"expand\")[0]\n",
    "\n",
    "# hm, hmkwargs = get_hm_double(clrM, clrP, _i1_5, _i2_5, **vlims_dict[\"50\"])\n",
    "# hmkwargs[\"interpolation\"] = \"gaussian\"\n",
    "# hmkwargs[\"interpolation_stage\"] = \"rgba\"\n",
    "# plt.imshow(hm, **hmkwargs)\n",
    "# ax = plt.gca()\n",
    "# # draw domains on the zoomed in heatmap ...\n",
    "# _big_boxes_kwargs = dict(lw=1., ec=\"green\", fc=\"none\", halo=100, ext_width=1)\n",
    "# for box in draw_ondiag_domains(_domains_region, (\"...\", (_i1_5, _i2_5), (_i1_5, _i2_5)), **_big_boxes_kwargs):\n",
    "#     ax.add_patch(box)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00ff8d6c-7d3a-413c-b214-061b7f7701d8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de501b1c-65a9-4076-b7ec-1f7cb782067f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a0a1903-33c6-4463-a2d9-c42813aff608",
   "metadata": {},
   "outputs": [],
   "source": [
    "_region5k = (\n",
    "    'chr8',\n",
    "    29_380_000,\n",
    "    40_000_000,\n",
    ")\n",
    "_region5k_zoom_A = (\n",
    "    'chr8',\n",
    "    29_500_000,\n",
    "    31_400_000,\n",
    ")\n",
    "_region5k_zoom_B = (\n",
    "    'chr8',\n",
    "    36_490_000,\n",
    "    39_830_000,\n",
    ")\n",
    "\n",
    "fracAB = (_region5k_zoom_B[2] - _region5k_zoom_B[1]) / (_region5k_zoom_A[2] - _region5k_zoom_A[1])\n",
    "\n",
    "_msample = \"m5hR1R2\"\n",
    "clrM = telo_clrs25[_msample]\n",
    "\n",
    "_psample = \"p5hR1R2\"\n",
    "clrP = telo_clrs25[_psample]\n",
    "\n",
    "clrM_zoom = telo_clrs10[_msample]\n",
    "clrP_zoom = telo_clrs10[_psample]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8e57d9d-24ff-4a69-81d3-e88476c07bca",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# and region1 and 2 @ 10kb ...\n",
    "_i1, _i2 = clrM.extent(_region5k)\n",
    "\n",
    "_i1_A, _i2_A = clrM_zoom.extent(_region5k_zoom_A)\n",
    "_i1_B, _i2_B = clrM_zoom.extent(_region5k_zoom_B)\n",
    "\n",
    "_i1_A_inset, _i2_A_inset = clrM.extent(_region5k_zoom_A)\n",
    "_i1_B_inset, _i2_B_inset = clrM.extent(_region5k_zoom_B)\n",
    "# ...\n",
    "_i1_A_inset -= _i1\n",
    "_i2_A_inset -= _i1\n",
    "# ...\n",
    "_i1_B_inset -= _i1\n",
    "_i2_B_inset -= _i1\n",
    "# ...\n",
    "size_A = _i2_A_inset-_i1_A_inset\n",
    "size_B = _i2_B_inset-_i1_B_inset\n",
    "\n",
    "patch_kwargs = dict(color=\"black\", fc=\"none\", lw=0.5)\n",
    "\n",
    "margin = 0.2\n",
    "matw = 1.333\n",
    "cbarh = 0.07\n",
    "cbarw = 0.7*matw\n",
    "mat_to_mcd = 0.125*margin\n",
    "chrom_to_mat = 0.3*margin\n",
    "zoom_mat = 0.475*matw\n",
    "\n",
    "\n",
    "_mev_color = \"tab:blue\"\n",
    "_pev_color = \"tab:red\"\n",
    "\n",
    "# The first items are for padding and the second items are for the axes, sizes are in inch.\n",
    "h = [\n",
    "    # 0,1,2,3\n",
    "    Size.Fixed(margin), Size.Fixed(matw*0.1), Size.Fixed(chrom_to_mat), Size.Fixed(matw),  # 25kb hic\n",
    "    Size.Fixed(mat_to_mcd), Size.Fixed(cbarh),  # 4,5\n",
    "    # 6\n",
    "    Size.Fixed(margin),\n",
    "    # split matrix into 4 pieces for zoomin\n",
    "    # 7\n",
    "    Size.Fixed(zoom_mat*fracAB),\n",
    "    Size.Fixed(mat_to_mcd), Size.Fixed(cbarh), Size.Fixed(margin) # vert MCDs\n",
    "]\n",
    "# goes from bottom to the top ...\n",
    "v = [\n",
    "    # 0\n",
    "    Size.Fixed(margin),\n",
    "    Size.Fixed(cbarh), Size.Fixed(0.2*margin),\n",
    "    # 1,2\n",
    "    Size.Fixed(matw*0.11), Size.Fixed(matw*0.075),  # genes, 27ac\n",
    "    # 3,4,5\n",
    "    Size.Fixed(matw*0.1), Size.Fixed(0.2*margin), Size.Fixed(cbarh),  # EVs, MCDs\n",
    "    # 6\n",
    "    Size.Fixed(mat_to_mcd),\n",
    "    # 7,8,9\n",
    "    # split matrix into 4 quadrants ...\n",
    "    Size.Fixed(zoom_mat), Size.Fixed(0.05*matw), Size.Fixed(zoom_mat),\n",
    "    Size.Fixed(margin),\n",
    "]\n",
    "\n",
    "# set figsize based on the tiling provided ...\n",
    "fig_width = sum(_h.fixed_size for _h in h)\n",
    "fig_height = sum(_v.fixed_size for _v in v)\n",
    "fig = plt.figure(\n",
    "    figsize=(fig_width, fig_height),\n",
    "    # facecolor='lightblue'\n",
    ")\n",
    "print(f\"figure size {fig_width=} {fig_height=}\")\n",
    "\n",
    "divider = Divider(fig, (0, 0, 1, 1), h, v, aspect=False)\n",
    "\n",
    "\n",
    "axs = {}\n",
    "# # 10 kb region1 and region2 ....\n",
    "axs[\"chrom\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1, ny=7+2, ny1=10+2))\n",
    "axs[\"hmap\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1+2, ny=7+2, ny1=10+2))\n",
    "\n",
    "axs[\"hmap_zoom_up\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=9+2))\n",
    "axs[\"hmap_zoom_bot\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=7+2))\n",
    "\n",
    "axs[\"mcd_zoom_up\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=5+4, ny=9+2))\n",
    "axs[\"mcd_zoom_bot\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=5+4, ny=7+2))\n",
    "\n",
    "\n",
    "# axs[\"hmap_zoom_C\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=7+2))\n",
    "# axs[\"hmap_zoom_D\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=5+4, ny=7+2))\n",
    "\n",
    "axs[\"genes\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1+2, ny=1+2))\n",
    "axs[\"k27ac\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1+2, ny=2+2))\n",
    "axs[\"evs\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1+2, ny=3+2))\n",
    "axs[\"mcd\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1+2, ny=5+2))\n",
    "axs[\"mcd_vert\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+2, ny=7+2, ny1=10+2))\n",
    "\n",
    "axs[\"genes_zoom\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=1+2))\n",
    "axs[\"k27ac_zoom\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=2+2))\n",
    "axs[\"evs_zoom\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=3+2))\n",
    "axs[\"mcd_zoom\"] = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=5+2))\n",
    "\n",
    "mtrack = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_msample}.{binsize25//1_000}kb.bw\",_region5k[0],_region5k[1],_region5k[2], bins=_i2-_i1 ))\n",
    "ptrack = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_psample}.{binsize25//1_000}kb.bw\",_region5k[0],_region5k[1],_region5k[2], bins=_i2-_i1 ))\n",
    "\n",
    "\n",
    "cbar = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=1+2, ny=1))\n",
    "cbar_zoom = fig.add_axes(divider.get_position(), axes_locator=divider.new_locator(nx=3+4, ny=1))\n",
    "\n",
    "\n",
    "# plot EV tracks ...\n",
    "axs[\"evs\"].axhline(0, lw=0.5, color=\"grey\")\n",
    "axs[\"evs\"].plot(mtrack, lw=0.7, color=_mev_color)\n",
    "axs[\"evs\"].plot(ptrack, lw=0.7, color=_pev_color)\n",
    "axs[\"evs\"].set_xlim(0, len(mtrack))\n",
    "\n",
    "\n",
    "# _region5k_zoom_B\n",
    "mtrack_zoom = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_msample}.{binsize10//1_000}kb.bw\",_region5k_zoom_B[0],_region5k_zoom_B[1],_region5k_zoom_B[2], bins=_i2_B - _i1_B ))\n",
    "ptrack_zoom = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_psample}.{binsize10//1_000}kb.bw\",_region5k_zoom_B[0],_region5k_zoom_B[1],_region5k_zoom_B[2], bins=_i2_B - _i1_B ))\n",
    "\n",
    "\n",
    "axs[\"evs_zoom\"].axhline(0, lw=0.5, color=\"grey\")\n",
    "axs[\"evs_zoom\"].plot(mtrack_zoom, lw=0.7, color=_mev_color)\n",
    "axs[\"evs_zoom\"].plot(ptrack_zoom, lw=0.7, color=_pev_color)\n",
    "axs[\"evs_zoom\"].set_xlim(0, len(mtrack_zoom))\n",
    "\n",
    "\n",
    "\n",
    "#################################\n",
    "# chrom schematics on both axes ...\n",
    "ax = axs[\"chrom\"]\n",
    "_8pl = arms_dict[\"chr8_p\"]\n",
    "_8ql = arms_dict[\"chr8_q\"]\n",
    "_regions_highlight = [(\"p\", _8pl-26, _8pl-38, \"dimgray\"), ]\n",
    "_ww = 0.1*(_8pl+_8ql)\n",
    "_8, _8hs = get_chrom_path_wpathes(\n",
    "    (0, 0),  # position of the centromere ...\n",
    "    parm=_8pl,  # p arm length\n",
    "    qarm=_8ql,  # q arm length\n",
    "    width=_ww,  # width\n",
    "    color=\"lightsteelblue\",\n",
    "    patches=_regions_highlight,\n",
    ")\n",
    "ax.set_axis_off()\n",
    "ax.add_artist(_8)\n",
    "for _h in _8hs:\n",
    "    ax.add_artist(_h)\n",
    "ax.set_xlim(-1.0*_ww, 1.0*_ww)\n",
    "ax.set_ylim(-(_8ql)-(0.11+0.1)*_ww, (_8pl)+(0.11+0.1)*_ww)\n",
    "##################################################\n",
    "ax_chrom = axs[\"chrom\"]\n",
    "ax = axs[\"hmap\"]\n",
    "con = ConnectionPatch(\n",
    "    xyA=(0.5*_ww, _8pl-26),\n",
    "    coordsA=ax_chrom.transData,\n",
    "    xyB=(0.0, 1.0),\n",
    "    coordsB=ax.transAxes,\n",
    "    **patch_kwargs,\n",
    ")\n",
    "ax.add_artist(con)\n",
    "con = ConnectionPatch(\n",
    "    xyA=(0.5*_ww, _8pl-38),\n",
    "    coordsA=ax_chrom.transData,\n",
    "    xyB=(0.0, 0.0),\n",
    "    coordsB=ax.transAxes,\n",
    "    **patch_kwargs,\n",
    ")\n",
    "ax.add_artist(con)\n",
    "# ##################################################\n",
    "\n",
    "\n",
    "# # _region5k_zoom_A\n",
    "# mtrack_zoom = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_msample}.{binsize10//1_000}kb.bw\",_region5k_zoom_B[0],_region5k_zoom_B[1],_region5k_zoom_B[2], bins=_i2_5z-_i1_5z ))\n",
    "# ptrack_zoom = np.nan_to_num(bbi.fetch(f\"ev_bigwig/{_psample}.{binsize10//1_000}kb.bw\",_region5k_zoom_B[0],_region5k_zoom_B[1],_region5k_zoom_B[2], bins=_i2_5z-_i1_5z ))\n",
    "\n",
    "# ########################\n",
    "# Plots genes !\n",
    "# region 1\n",
    "_start, _end = _region5k[1], _region5k[2]\n",
    "_ax = axs[\"genes\"]\n",
    "#\n",
    "_gfs_fwd = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **fwd_inact_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_fwd, _region5k), pad=7_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_rev = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **rev_inact_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_rev, _region5k), pad=7_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_fwd.plot(ax=_ax, x_lim=(_start, _end), level_offset=-0.45, **plotargs)\n",
    "_gfs_rev.plot(ax=_ax, x_lim=(_start, _end), level_offset=0.45, **plotargs)\n",
    "#\n",
    "_gfs_fwd = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **fwd_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_fwd_act, _region5k), pad=7_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_rev = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **rev_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_rev_act, _region5k), pad=7_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_fwd.plot(ax=_ax, x_lim=(_start, _end), level_offset=-0.45, **plotargs)\n",
    "_gfs_rev.plot(ax=_ax, x_lim=(_start, _end), level_offset=0.45, **plotargs)\n",
    "_ax.set_ylim(-1,1)\n",
    "_ax.set_axis_on()\n",
    "\n",
    "\n",
    "_start, _end = _region5k_zoom_B[1], _region5k_zoom_B[2]\n",
    "_ax = axs[\"genes_zoom\"]\n",
    "_gfs_fwd = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **fwd_inact_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_fwd, _region5k_zoom_B), pad=7_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_rev = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **rev_inact_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_rev, _region5k_zoom_B), pad=7_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_fwd.plot(ax=_ax, x_lim=(_start, _end), level_offset=-0.45, **plotargs)\n",
    "_gfs_rev.plot(ax=_ax, x_lim=(_start, _end), level_offset=0.45, **plotargs)\n",
    "# active noiw ...\n",
    "_gfs_fwd = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **fwd_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_fwd_act, _region5k_zoom_B), pad=7_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_rev = GraphicRecord(\n",
    "    sequence_length=(_end-_start+1),\n",
    "    features=[\n",
    "        GraphicFeature(_row.start, _row.end, **rev_args) for _row in \\\n",
    "        bioframe.merge(\n",
    "            bioframe.expand(bioframe.select(_rev_act, _region5k_zoom_B), pad=7_500),\n",
    "            min_dist = 20_000,\n",
    "        ).itertuples(index=False)\n",
    "    ],\n",
    ")\n",
    "_gfs_fwd.plot(ax=_ax, x_lim=(_start, _end), level_offset=-0.45, **plotargs)\n",
    "_gfs_rev.plot(ax=_ax, x_lim=(_start, _end), level_offset=0.45, **plotargs)\n",
    "_ax.set_ylim(-1,1)\n",
    "_ax.set_axis_on()\n",
    "\n",
    "\n",
    "########################################\n",
    "for k, ax in axs.items():\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "\n",
    "hm, hmkwargs = get_hm_double(clrM, clrP, _i1, _i2, **vlims_dict[\"50\"])\n",
    "hmkwargs[\"interpolation\"] = \"gaussian\"\n",
    "hmkwargs[\"interpolation_stage\"] = \"rgba\"\n",
    "axs[\"hmap\"].imshow(hm, **hmkwargs)\n",
    "\n",
    "\n",
    "# _i1_A_inset, _i2_A_inset\n",
    "# _i1_B_inset, _i2_B_inset\n",
    "_rec = Rectangle(\n",
    "    (_i1_A_inset, _i1_B_inset),\n",
    "    size_A, size_B,\n",
    "    **patch_kwargs,\n",
    ")\n",
    "# draw those box and lines ...\n",
    "axs[\"hmap\"].add_patch(_rec)\n",
    "\n",
    "\n",
    "# _i1_A_inset, _i2_A_inset\n",
    "# _i1_B_inset, _i2_B_inset\n",
    "_rec = Rectangle(\n",
    "    (_i1_B_inset, _i1_A_inset),\n",
    "    size_B, size_A,\n",
    "    **patch_kwargs,\n",
    ")\n",
    "# draw those box and lines ...\n",
    "axs[\"hmap\"].add_patch(_rec)\n",
    "\n",
    "\n",
    "vlims_dict[\"10\"] = dict(vmin=0.00005, vmax=0.002)\n",
    "hm, hmkwargs = get_hm(clrM_zoom, _i1_A, _i2_A, _i1_B, _i2_B, **vlims_dict[\"10\"])\n",
    "hmkwargs[\"interpolation\"] = \"antialiased\"\n",
    "hmkwargs[\"interpolation_stage\"] = \"data\"\n",
    "axs[\"hmap_zoom_up\"].imshow(hm, **hmkwargs)\n",
    "\n",
    "hm, hmkwargs = get_hm(clrP_zoom, _i1_A, _i2_A, _i1_B, _i2_B, **vlims_dict[\"10\"])\n",
    "hmkwargs[\"interpolation\"] = \"antialiased\"\n",
    "hmkwargs[\"interpolation_stage\"] = \"data\"\n",
    "axs[\"hmap_zoom_bot\"].imshow(hm, **hmkwargs)\n",
    "# _i1_zA, _i2_zA\n",
    "# _i1_zB, _i2_zB\n",
    "\n",
    "\n",
    "_track_factor=10_000\n",
    "_bin_factor = 500/(_region5k_zoom_B[2] - _region5k_zoom_B[1])\n",
    "#do tracks in a cycle ...\n",
    "_track, _bin_factor, _vlims, _ax1, _ax2, _color = (\n",
    "    bws['H3K27ac'],\n",
    "    2*_bin_factor*0.75,\n",
    "    (0.001, 75),\n",
    "    axs[\"k27ac\"],\n",
    "    axs[\"k27ac_zoom\"],\n",
    "    \"red\",\n",
    ")\n",
    "\n",
    "_bins1 = int( _bin_factor*( _region5k[2] - _region5k[1] ) )\n",
    "_bins2 = int( _bin_factor*( _region5k_zoom_B[2] - _region5k_zoom_B[1] ) )\n",
    "track = bbi.fetch(_track, _region5k[0], _region5k[1], _region5k[2], bins=_bins1, summary=\"mean\", missing=0.0, oob=0.0)\n",
    "track_zoom_B = bbi.fetch(_track, _region5k_zoom_B[0], _region5k_zoom_B[1], _region5k_zoom_B[2], bins=_bins2, summary=\"mean\", missing=0.0, oob=0.0)\n",
    "_ax1.fill_between(\n",
    "    x=np.arange(len(track)),\n",
    "    y1=0,\n",
    "    y2=track,\n",
    "    linewidth=0.5,\n",
    "    color=_color,\n",
    ")\n",
    "_ax1.set_xlim(0, _bins1)\n",
    "_ax1.set_ylim(*_vlims)\n",
    "_ax2.fill_between(\n",
    "    x=np.arange(len(track_zoom_B)),\n",
    "    y1=0,\n",
    "    y2=track_zoom_B,\n",
    "    linewidth=0.5,\n",
    "    color=_color,\n",
    ")\n",
    "_ax2.set_xlim(0, _bins2)\n",
    "_ax2.set_ylim(*_vlims)\n",
    "\n",
    "# plot MCDs ...\n",
    "mcd_kwargs = dict(color=\"black\", alpha=1,ec=None)\n",
    "# doing that using clr.offset and clr.extent functionality:\n",
    "_anchors[\"bin_id\"] = _anchors[['chrom', 'peak_start', 'peak_end']].apply(clrM.offset, axis=1, result_type=\"expand\")\n",
    "_anchors[\"bin_width\"] = _anchors[['chrom', 'peak_start', 'peak_end']] \\\n",
    "    .apply(clrM.extent, axis=1, result_type=\"expand\") \\\n",
    "    .apply(np.diff, axis=1, result_type=\"expand\")[0]\n",
    "rectangles_at_anchors(_anchors, clrM.extent(_region5k), axs[\"mcd\"], ext_width=0.0, **mcd_kwargs)\n",
    "axs[\"mcd\"].set_xlim(0, _i2-_i1)\n",
    "\n",
    "# vertical ones ...\n",
    "rectangles_at_anchors_vert(_anchors, clrM.extent(_region5k), axs[\"mcd_vert\"], ext_width=0.0, **mcd_kwargs)\n",
    "axs[\"mcd_vert\"].set_ylim(0, _i2-_i1)\n",
    "axs[\"mcd_vert\"].invert_yaxis()\n",
    "\n",
    "# doing that using clr.offset and clr.extent functionality:\n",
    "_anchors[\"bin_id\"] = _anchors[['chrom', 'peak_start', 'peak_end']].apply(clrM_zoom.offset, axis=1, result_type=\"expand\")\n",
    "_anchors[\"bin_width\"] = _anchors[['chrom', 'peak_start', 'peak_end']] \\\n",
    "    .apply(clrM_zoom.extent, axis=1, result_type=\"expand\") \\\n",
    "    .apply(np.diff, axis=1, result_type=\"expand\")[0]\n",
    "rectangles_at_anchors(_anchors, clrM_zoom.extent(_region5k_zoom_B), axs[\"mcd_zoom\"], ext_width=0.0, **mcd_kwargs)\n",
    "axs[\"mcd_zoom\"].set_xlim(0, _i2_B-_i1_B)\n",
    "\n",
    "\n",
    "# vertical ones ...\n",
    "rectangles_at_anchors_vert(_anchors, clrM_zoom.extent(_region5k_zoom_A), axs[\"mcd_zoom_up\"], ext_width=0.0, **mcd_kwargs)\n",
    "axs[\"mcd_zoom_up\"].set_ylim(0, _i2_A-_i1_A)\n",
    "axs[\"mcd_zoom_up\"].invert_yaxis()\n",
    "rectangles_at_anchors_vert(_anchors, clrM_zoom.extent(_region5k_zoom_A), axs[\"mcd_zoom_bot\"], ext_width=0.0, **mcd_kwargs)\n",
    "axs[\"mcd_zoom_bot\"].set_ylim(0, _i2_A-_i1_A)\n",
    "axs[\"mcd_zoom_bot\"].invert_yaxis()\n",
    "\n",
    "axs[\"genes\"].axis(\"off\")\n",
    "axs[\"k27ac\"].axis(\"off\")\n",
    "axs[\"evs\"].axis(\"off\")\n",
    "\n",
    "axs[\"genes_zoom\"].axis(\"off\")\n",
    "axs[\"k27ac_zoom\"].axis(\"off\")\n",
    "axs[\"evs_zoom\"].axis(\"off\")\n",
    "\n",
    "\n",
    "\n",
    "# it's 25 actually - but whatever\n",
    "_vmin, _vmax = vlims_dict[\"50\"][\"vmin\"], vlims_dict[\"50\"][\"vmax\"]\n",
    "plt.colorbar(\n",
    "    cm.ScalarMappable(norm=LogNorm(vmin=_vmin, vmax=_vmax), cmap=mycmap),\n",
    "    cax=cbar,\n",
    "    orientation=\"horizontal\",\n",
    ")\n",
    "cbar.xaxis.set_visible(True)\n",
    "_ticks_power = range(\n",
    "    np.ceil(np.log10(_vmin)).astype(int),\n",
    "    np.floor(np.log10(_vmax)).astype(int)+1,\n",
    ")\n",
    "cbar.set_xticks([np.power(10.,_p) for _p in _ticks_power])\n",
    "cbar.set_xticklabels(['$\\mathdefault{10^{%i}}$' % l for l in _ticks_power])\n",
    "cbar.xaxis.set_tick_params(labelsize=6, pad=0.1, length=2)\n",
    "cbar.minorticks_off()\n",
    "\n",
    "\n",
    "# it's 25 actually - but whatever\n",
    "_vmin, _vmax = vlims_dict[\"10\"][\"vmin\"], vlims_dict[\"10\"][\"vmax\"]\n",
    "plt.colorbar(\n",
    "    cm.ScalarMappable(norm=LogNorm(vmin=_vmin, vmax=_vmax), cmap=mycmap),\n",
    "    cax=cbar_zoom,\n",
    "    orientation=\"horizontal\",\n",
    ")\n",
    "cbar_zoom.xaxis.set_visible(True)\n",
    "_ticks_power = range(\n",
    "    np.ceil(np.log10(_vmin)).astype(int),\n",
    "    np.floor(np.log10(_vmax)).astype(int)+1,\n",
    ")\n",
    "cbar_zoom.set_xticks([np.power(10.,_p) for _p in _ticks_power])\n",
    "cbar_zoom.set_xticklabels(['$\\mathdefault{10^{%i}}$' % l for l in _ticks_power])\n",
    "cbar_zoom.xaxis.set_tick_params(labelsize=6, pad=0.1, length=2)\n",
    "cbar_zoom.minorticks_off()\n",
    "\n",
    "\n",
    "plt.savefig(\"Fig5K.svg\", dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "013805fb-7e12-4b8b-bc5a-ec2205030c8c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "638d9b3f-96d4-485c-89af-494b530f1852",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bc88e0f-0990-4e73-937b-da9eb29c11f2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd690125-9194-4b10-aae9-e2cb4e4b8325",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b4a7d54-a38d-49e4-a523-3b79531d9c59",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
